---
title: "stackr missing data visualization analysis"
author: "Thierry Gosselin"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
# output: 
# pdf_document:
vignette: >
  %\VignetteIndexEntry{"stackr missing data visualization analysis"}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---


# Objectives
Learn how to visualize missing genotypes in your genomic dataset with 
the function `stackr::missing_visualization` (time = 15 min).


# Workflow
The function `missing_visualization` in 
[stackr](https://github.com/thierrygosselin/stackr) uses various genomic input 
files and conduct identity-by-missingness analyses (IBM) using 
Principal Coordinates Analysis (PCoA), also called Multidimensional Scaling (MDS). 
Figures and summary tables of missing information at the marker, 
individual and population level are generated. 
Below, the simplest form of the function to get results ... fast! 
More options are available, see function documentation.


## Prepare your R workspace
1. Clean your desk and remove the clutter in the global environment
```r
rm(list=ls())
```

1. Load the required libraries (copy/paste in one block):
```r
rm(list=ls())
if (!require("pacman")) install.packages("pacman")
library("pacman")
pacman::p_load(devtools, reshape2, ggplot2, stringr, stringi, plyr, dplyr, tidyr, readr, purrr, data.table, ape, adegenet, parallel, lazyeval, randomForestSRC)
if (!require("stackr")){
  install_github("thierrygosselin/stackr", build_vignettes = TRUE)
  library("stackr")
}
```

1. Set your working directory
```
setwd("~/Documents")
```

## Get the data and create a strata file
1. Dataset: in this example, we use the data in Ferchaud and Hansen (2015 and 2016) 
paper.
```r
writeBin(httr::content(httr::GET("http://datadryad.org/bitstream/handle/10255/dryad.97237/sticklebacks_Danish.vcf?sequence=1"), "raw"), "example_vcf2dadi_ferchaud_2015.vcf")
```
1. We need to generate a strata file necessary to make the vcf population-wise. 
The strata file is a tab-delimited file or object, in the global environment,
with a minimum of 2 columns: `INDIVIDUALS` and `STRATA`. 
The `STRATA` and remaining columns can be any hierarchical grouping.
    * Because the population id is contained in the name of the individuals and
  easily extracted with string command, we use the 
  function `stackr::individuals2strata`. 
    ```r
    # here is one individual id: 
    Hadsten10_q_75.align
    # with pop starting at position 1 and to keep only 3 characters (necessary to discriminate populations from each other) we stop at... 3
      ```
  
    * For info on the function use: `?individuals2strata`.
  Don't have this kind of names... you can also make the strata file by hand, 
  the old fashion way.

    * For your data, you might not need to make a strata file if you have already
  run \href{http://catchenlab.life.illinois.edu/stacks/}{stacks} on your data, 
the strata file is similar to a stacks `population map file`, make sure you 
have the required column names  (`INDIVIDUALS` and `STRATA`). 

    * We will extract the id from the first line of the vcf file. 
  If the individuals were in a single column in a file (e.g. "individuals.txt"), 
  make sure the column is named `INDIVIDUALS` and use the file directly 
  in the function `stackr::individuals2strata`.
  
    * We could put all the lines below in one call, but for clarity we'll break down
the steps:
    ```r
# Isolate the first line
id.vcf <- data.table::fread(
  input = "example_vcf2dadi_ferchaud_2015.vcf",
  sep = "\t",
  stringsAsFactors = FALSE, 
  header = TRUE,
  drop = c("POS", "ID", "REF", "ALT", "QUAL", "FILTER", "INFO", "FORMAT"),
  skip = "CHROM",
  nrows = 1,
  showProgress = TRUE,
  verbose = FALSE
) %>% 
  as_data_frame() %>% 
  select(-`#CHROM`)
  # to keep only the column name
  individuals <- data.frame(INDIVIDUALS = colnames(id.vcf)) 
    ```
    ```r
# create the strata
strata.stickleback <- stackr::individuals2strata(
data = individuals, 
strata.start = 1,
strata.end = 3,
filename = "strata.stickleback.tsv"
)
    ```

##Run the function stackr::missing_visualization
```r
ibm.stickleback <- stackr::missing_visualization(
data = "example_vcf2dadi_ferchaud_2015.vcf", # VCF file, but see function documentation for other options
strata = "strata.stickleback.tsv") # the strata file just downloaded
```

This is the simplest way to run the function. And you should see this in RStudio console:

```r
#######################################################################
################### stackr: missing_visualization #####################
#######################################################################
Using input file in your directory
File type: VCF
Whitelist of markers: no
Blacklisted individuals: no
strata file: yes
Importing the VCF...
Making the VCF population wise
Tidying the vcf...
Erasing genotype: no
Removing monomorphic markers: yes
Scanning for monomorphic markers...
Number of monomorphic markers removed: 0
Including the strata file
Principal Coordinate Analysis (PCoA)...
Generating Identity by missingness plot
Generating missing information summary tables and plot
Missingness per individuals
Missingness per populations
Missingness per markers
```

A new object `ibm.stickleback` was created in your global environment. 
It's a list and to view it's content use `names(ibm.stickleback)`

```r
"tidy.data"
"ibm.plot.strata.POP_ID"
"blacklist.id.missing.10"
"blacklist.id.missing.20"
"blacklist.id.missing.30"
"blacklist.id.missing.40"
"blacklist.id.missing.50"
"blacklist.id.missing.60"
"blacklist.id.missing.70"
"vectors"
"heatmap"
"missing.genotypes.ind"
"missing.genotypes.ind.plot"
"missing.genotypes.pop"
"missing.genotypes.pop.plot"
"missing.genotypes.markers"
"missing.genotypes.markers.plot"
```
The function generates by default the IBM-PCoA plot, the heatmap plot and several plots showing the 
distribution of missing genotypes by individuals, pop and markers. Blacklists of individuals
are included in the list `ibm.stickleback` and written to the working directory (defaults from 10% to 70%, that automatically stops if no individuals at the threshold is found, can be changed with `ind.missing.geno.threshold` argument). Tables with summary missing information along the vectors of eigenvalues of the principal coordinates analysis are also part of the list output. All these objects are accessed with `$`, see the next step.

### Visualization

```r
# To view the IBM-PCoA plot:
ibm.stickleback$ibm.plot.strata.POP_ID

# to put the heatmap figure in a new object, outside the list
heatmap <- ibm.stickleback$heatmap 

# To view the heatmap of missing genotypes per markers and individuals:
heatmap

# To save the plots in a vectorial pdf:
ggsave(filename = "stickleback.missing.heatmap.pdf", width = 10, height = 20, dpi = 300, units = "cm", useDingbats = F)

# To save the plots in a vectorial png:
ggsave(filename = "stickleback.missing.heatmap.png", width = 10, height = 20, dpi = 100, units = "cm")

# To view the distribution of missing genotypes per individuals
ibm.stickleback$missing.genotypes.ind.plot

# To view the distribution of missing genotypes per markers and populations
ibm.stickleback$missing.genotypes.pop.plot
```

**The heatmap should look like this:**

![heatmap](https://raw.githubusercontent.com/thierrygosselin/stackr/master/vignettes/stickleback.missing.heatmap.png)

**The IBM-PCoA plot:**

![IBM-PCoA plot](https://raw.githubusercontent.com/thierrygosselin/stackr/master/vignettes/stickleback.ibm.png)


**Table summary with missing info per individuals**: `table.ind <- ibm.stickleback$missing.genotypes.ind`

|INDIVIDUALS            |POP_ID | MISSING_GENOTYPE| MARKER_NUMBER|  PERC|
|:----------------------|:------|----------------:|-------------:|-----:|
|Hadsten1-q-75.align    |Had    |             3887|         31802| 12.22|
|Hadsten10-q-75.align   |Had    |              125|         31802|  0.39|
|Hadsten11-q-75.align   |Had    |               55|         31802|  0.17|
|Hadsten12-q-75.align   |Had    |               90|         31802|  0.28|
|Hadsten13-q-75.align   |Had    |               77|         31802|  0.24|
|Hadsten14-q-75.align   |Had    |              287|         31802|  0.90|


**The boxplot of missing per populations:** `boxplot.missing.pop <- ibm.stickleback$missing.genotypes.pop.plot`

![boxplot](https://raw.githubusercontent.com/thierrygosselin/stackr/master/vignettes/stickleback.missing.boxplot.pop.png)



###Filtering strategies
* use the blacklist of individuals created with the function to **manage** missing data inside filtering pipeline
* excluding/including individuals, populations, markers to test the impact of filtering on polymorphism discovery
* ready to test missing data imputations ? use most of stackr functions with the imputations arguments...

##References

Ferchaud A, Hansen MM (2016) The impact of selection, gene flow and demographic history on heterogeneous genomic divergence: threespine sticklebacks in divergent environments. Molecular Ecology 25(1): 238–259. http://dx.doi.org/10.1111/mec.13399

Ferchaud A, Hansen MM (2015) Data from: The impact of selection, gene flow and demographic history on heterogeneous genomic divergence: threespine sticklebacks in divergent environments. Dryad Digital Repository. http://dx.doi.org/10.5061/dryad.kp11q

Danecek P, Auton A, Abecasis G et al. (2011)
The variant call format and VCFtools. Bioinformatics, 27, 2156–2158.

Purcell S, Neale B, Todd-Brown K et al. (2007)
PLINK: a tool set for whole-genome association and population-based linkage
analyses. American Journal of Human Genetics, 81, 559–575.
