<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using stackr to run stacks pipeline inside R/RStudio • stackr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using stackr to run stacks pipeline inside R/RStudio">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stackr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/stackr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/thierrygosselin/stackr" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using stackr to run stacks pipeline inside
R/RStudio</h1>
                        <h4 data-toc-skip class="author">Thierry
Gosselin</h4>
            
            <h4 data-toc-skip class="date">2025-04-02</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/stackr/blob/HEAD/vignettes/stackr.Rmd" class="external-link"><code>vignettes/stackr.Rmd</code></a></small>
      <div class="hidden name"><code>stackr.Rmd</code></div>

    </div>

    
    
<p>This vignette purposes is to show users how to use stackr to run the
<a href="http://catchenlab.life.illinois.edu/stacks/" class="external-link">stacks</a>
pipeline inside R/RStudio.</p>
<div class="section level2">
<h2 id="stackr-assumptions">stackr assumptions<a class="anchor" aria-label="anchor" href="#stackr-assumptions"></a>
</h2>
<ul>
<li>stackr is similar to stacks concerning data quality: it won’t make
your sequencing data better. Bad input data is just that… bad!</li>
<li>You have read <a href="http://catchenlab.life.illinois.edu/stacks/manual/" class="external-link">stacks
manual</a> and understand the separate modules and the outputs that
stacks produces.</li>
<li>You have read the latest <a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/mec.15253" class="external-link">stacks
paper</a> (version 2, Rochette et al. 2019):</li>
</ul>
</div>
<div class="section level2">
<h2 id="stackr-advantages">stackr advantages<a class="anchor" aria-label="anchor" href="#stackr-advantages"></a>
</h2>
<ul>
<li>Run stacks inside R/RStudio/RStudio Server. If you don’t like R,
stackr is not the tool for you.</li>
<li>The philosophy of working by project with pre-organized
folders.</li>
<li>
<strong>stackr is not slowing down stacks</strong> it’s actually
making some steps faster with parallelization.</li>
<li>If you have &gt; 1 sequencing chip/lane: this workflow will save you
a lot of time.</li>
<li>
<strong>Technical replicates</strong>: inside or across chip/lanes
are managed uniquely (you want technical replicates…).</li>
<li>Noise reduction.</li>
<li>Data normalization.</li>
<li>
<strong>Crashing computer/cluster/server ?</strong>:
<ul>
<li>stackr is also on my <a href="https://thierrygosselin.github.io/RADseq_cloud_tutorial/radseq_cloud_tutorial.html" class="external-link">Amazon
AMI</a>
</li>
<li>no more nightmares of having to restart your stacks pipeline because
of a crashed server… stackr manage stacks unique integer (previously
called SQL IDs) throughout the pipeline. It’s integrated from the start,
making it a breeze to just re-start your pipeline after a crash!</li>
</ul>
</li>
<li>
<strong>mismatch testing:</strong> <em>de novo</em> mismatch
threshold series is integrated inside <code>run_ustacks</code> and
stackr will produce tables and figures automatically.</li>
<li>
<strong>catalog</strong>: for bigger sampling size project, breaking
down the catalog into several separate <em>cstacks</em> steps makes the
pipeline more rigorous if your computer/cluster/server crash.</li>
<li>
<strong>logs</strong> generated by stacks are read and transferred
in human-readable tables/tibbles. Detecting problems is easier.</li>
<li>summary of different stacks modules: available automatically inside
stackr pipeline, but also available for users who didn’t use stackr to
run stacks.</li>
<li>Detecting problems is easier.</li>
<li>Increased reproducibly.</li>
</ul>
</div>
<div class="section level2">
<h2 id="coding-style">Coding style<a class="anchor" aria-label="anchor" href="#coding-style"></a>
</h2>
<ul>
<li>stackr is trying hard to follow the <a href="https://style.tidyverse.org" class="external-link">tidyverse style guide</a>.</li>
<li>inside stacks, parallel execution and the use of threads sometimes
uses the argument <code>-p</code>, <code>-t</code>,
<code>--threads</code>. Depending on module the meaning changes, which
can be confusing or generate problems for first-time users. Inside
stackr, using parallel execution is always through the use of the
argument: <code>parallel.core</code>.</li>
<li>argument break down inside stacks modules sometimes uses the hyphen
<code>-</code>. Inside stackr this is replaced by function arguments
with <code>.</code> point/dot. Example: in <strong>stacks</strong>
<code>process_radtags</code> argument <code>--adapter-1</code> will be
the argument <code>adapter.1</code> inside <strong>stackr</strong>
<code>run_process_radtags</code>.</li>
</ul>
</div>
<div class="section level2">
<h2 id="workflow">Workflow<a class="anchor" aria-label="anchor" href="#workflow"></a>
</h2>
<p>The names of stackr functions follows stacks module, below the
similarity:</p>
<p><img src="stackr_workflow.png"></p>
<p>They are tons of other workflows and below are just a few
examples:</p>
<ul>
<li><a href="https://github.com/enormandeau/stacks_workflow" class="external-link">Louis
Bernatchez lab workflow using stacks</a></li>
<li><a href="http://www.ddocent.com" class="external-link">Jon Puritz’s dDocent
workflow</a></li>
<li><a href="https://ipyrad.readthedocs.io/en/latest/" class="external-link">Deren Eaton
ipyrad</a></li>
</ul>
<div class="section level3">
<h3 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h3>
<ul>
<li>Make sure you have the latest R and RStudio version before
installing stackr. My other package <a href="https://thierrygosselin.github.io/radiator/index.html" class="external-link">radiator</a>
as a nice <a href="https://thierrygosselin.github.io/radiator/articles/rad_genomics_computer_setup.html" class="external-link">vignette</a>
for this.</li>
<li>Install <a href="http://catchenlab.life.illinois.edu/stacks/manual/#install" class="external-link">stacks</a>
and it’s dependencies (my own installation pipeline <a href="https://thierrygosselin.github.io/RADseq_cloud_tutorial/radseq_cloud_tutorial.html#stacks" class="external-link">here</a>).</li>
<li><strong>don’t write to me if you’re having problem at this
step.</strong></li>
<li>
<a href="https://groups.google.com/forum/?fromgroups#!forum/stacks-users" class="external-link">stacks
google group</a> is your best friend.</li>
<li>Install stackr by following the <a href="https://github.com/thierrygosselin/stackr#installation" class="external-link">instructions</a>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="generate-project-folders">Generate project folders<a class="anchor" aria-label="anchor" href="#generate-project-folders"></a>
</h3>
<p>Here, I’m going to use RStudio, because I really got to like it and
I’m very efficient in it. But it’s by no mean essential.</p>
<ol style="list-style-type: decimal">
<li><p>Set your working directory inside RStudio using
<code>setwd</code> function and appropriate path.</p></li>
<li><p>Generate the folders for your project</p></li>
</ol>
<p>To get the doc on the function we’re about to use:
<code><a href="../reference/build_stackr_workflow_dir.html">?stackr::build_stackr_workflow_dir</a></code>.</p>
<p>Let’s start a project on turtles…</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/build_stackr_workflow_dir.html">build_stackr_workflow_dir</a></span><span class="op">(</span>main.folder.name <span class="op">=</span> <span class="st">"turtle"</span><span class="op">)</span></span>
<span><span class="co"># stackr::build_stackr_workflow_dir(main.folder.name = "turtle", date = FALSE) #if you don't want the date appended</span></span></code></pre></div>
<p>By default, the date and time is appended to the main folder and
inside, these folders will be generated:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="dv">01</span>_scripts</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="dv">02</span>_project_info</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="dv">03</span>_sequencing_lanes</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="dv">04</span>_process_radtags</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="dv">05</span>_clustering_mismatches</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="dv">06</span>_ustacks_2_gstacks</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="dv">07</span>_populations</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="dv">08</span>_stacks_results</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="dv">09</span>_log_files</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="dv">10</span>_filters</span></code></pre></div>
<p>For now let’s just say that you put the scripts and other
<code>.R</code> files associated to this project inside the first
folder: <code>01_scripts</code>.</p>
</div>
<div class="section level3">
<h3 id="start-a-project">Start a project<a class="anchor" aria-label="anchor" href="#start-a-project"></a>
</h3>
<p>Inside RStudio generate a new project inside the directory just
created:</p>
<p><img src="rstudio.project.1.png"></p>
<p><img src="rstudio.project.2.png"></p>
<p>You can add the alias/shorcut for the project on your computer
desktop.</p>
</div>
<div class="section level3">
<h3 id="project-info-file">Project info file<a class="anchor" aria-label="anchor" href="#project-info-file"></a>
</h3>
<p>Inside MS Excel or another software that works with tabular columns,
prepare your project info file and save it with the file extension
<code>.tsv</code> inside the second folder:
<code>02_project_info</code>. For example, the file
<code>project.info.turtle.tsv</code> will contain a minimum of 3 columns
for <strong>single-end</strong> and 4 columns for
<strong>paired-end</strong> sequencing. The number of lines equal the
number of sample: 1 sample per line. The project info file is
<strong>sample</strong> oriented or focused.</p>
<p>But before generating and filling the project info file consider the
names of your samples…</p>
<div class="section level4">
<h4 id="the-art-of-naming-your-samples">The art of naming your samples<a class="anchor" aria-label="anchor" href="#the-art-of-naming-your-samples"></a>
</h4>
<p>You already have somewhere on your computer barcodes and sample name.
Here is the time to reconsider how you named your sample. This is not
trivial, researcher loose a lot of time because of bad naming scheme.
REALLY.</p>
<p><strong>My philosophy:</strong></p>
<p>Include useful metadata associated with your sample separated by an
<a href="https://en.wikipedia.org/wiki/Hyphen" class="external-link">hyphen</a>
<code>-</code>. Why an hyphen <code>-</code> instead of a point
<code>.</code> or an underscore <code>_</code> ?? Imagine you print your
file on real paper or that you have to translate by hand the sample
names. An hyphen, when printed, generated far less confusion because it
can easily be distinguised if printed on paper (lined paper or not). The
underscore or point will just be confusion with space.</p>
<ul>
<li>Please don’t use space</li>
<li>Please use ALL CAPS so that confusion is minimized.</li>
<li>The key here is consistency.</li>
</ul>
<p>Below an example using metadata: <code>SPECIES</code>,
<code>STRATA</code> (could be location, I don’t know the pop yet),
<code>YEAR</code> (of sampling), <code>SEX</code>, <code>ID</code> (with
sufficient digits)</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">STU</span><span class="op">-</span><span class="va">BUR</span><span class="op">-</span><span class="fl">2019</span><span class="op">-</span><span class="cn">F</span><span class="op">-</span><span class="fl">0001</span></span>
<span><span class="va">STU</span><span class="op">-</span><span class="va">BUR</span><span class="op">-</span><span class="fl">2019</span><span class="op">-</span><span class="cn">F</span><span class="op">-</span><span class="fl">0002</span></span>
<span><span class="va">STU</span><span class="op">-</span><span class="va">BUR</span><span class="op">-</span><span class="fl">2019</span><span class="op">-</span><span class="cn">F</span><span class="op">-</span><span class="fl">0003</span></span>
<span><span class="va">STU</span><span class="op">-</span><span class="va">BUR</span><span class="op">-</span><span class="fl">2019</span><span class="op">-</span><span class="cn">F</span><span class="op">-</span><span class="fl">0004</span></span>
<span><span class="va">STU</span><span class="op">-</span><span class="va">BUR</span><span class="op">-</span><span class="fl">2019</span><span class="op">-</span><span class="va">M</span><span class="op">-</span><span class="fl">0005</span></span>
<span><span class="va">STU</span><span class="op">-</span><span class="va">BUR</span><span class="op">-</span><span class="fl">2017</span><span class="op">-</span><span class="cn">F</span><span class="op">-</span><span class="fl">0006</span></span>
<span><span class="va">STU</span><span class="op">-</span><span class="va">GRA</span><span class="op">-</span><span class="fl">2017</span><span class="op">-</span><span class="va">M</span><span class="op">-</span><span class="fl">0007</span> <span class="co"># you could start over from 0001 with different strata, it doesn't matter</span></span>
<span><span class="va">STU</span><span class="op">-</span><span class="va">GRA</span><span class="op">-</span><span class="fl">2018</span><span class="op">-</span><span class="va">M</span><span class="op">-</span><span class="fl">0008</span></span>
<span><span class="va">STU</span><span class="op">-</span><span class="va">GRA</span><span class="op">-</span><span class="fl">2019</span><span class="op">-</span><span class="va">M</span><span class="op">-</span><span class="fl">0009</span></span>
<span><span class="va">STU</span><span class="op">-</span><span class="va">GRA</span><span class="op">-</span><span class="fl">2019</span><span class="op">-</span><span class="cn">F</span><span class="op">-</span><span class="fl">0010</span></span>
<span><span class="va">STU</span><span class="op">-</span><span class="va">GRA</span><span class="op">-</span><span class="fl">2019</span><span class="op">-</span><span class="cn">F</span><span class="op">-</span><span class="fl">0011</span></span></code></pre></div>
<p><strong>Special considerations:</strong></p>
<ul>
<li>keep it simple.</li>
<li>you can use 3 or 4 letters for the metadata, but inside 1 metadata
but please please <strong>be consistent</strong>.</li>
<li>future parsing, if you need it, will be very very easy this
way.</li>
</ul>
<p>Ok now the project info file…</p>
</div>
<div class="section level4">
<h4 id="single-end">single-end<a class="anchor" aria-label="anchor" href="#single-end"></a>
</h4>
<p><strong>Columns prerequisites:</strong></p>
<ul>
<li>
<code>LANES</code>: the name of the sequencing file
e.g. H6CY7BBXY_3_254.fastq.gz</li>
<li>
<code>BARCODES</code>: the barcode associated with the sample.</li>
<li>
<code>INDIVIDUALS</code>: a more appealing sample name (more on this
below).</li>
</ul>
<p>Generate an empty file like this:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  LANES <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, </span>
<span>  BARCODES <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, </span>
<span>  INDIVIDUALS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_tsv</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.</span>, path <span class="op">=</span> <span class="st">"02_project_info/project.info.turtle.tsv"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="paired-end">paired-end<a class="anchor" aria-label="anchor" href="#paired-end"></a>
</h4>
<p><strong>Columns prerequisites:</strong></p>
<ul>
<li>
<code>BARCODES</code>: the barcode associated with the sample</li>
<li>
<code>INDIVIDUALS</code>: a more appealing sample name (more on this
below)</li>
<li>
<code>FORWARD</code>: the name of the forward sequencing file
e.g. H5CY5BBXY_2_25_1.fastq.gz</li>
<li>
<code>REVERSE</code>: the name of the reverse sequencing file
e.g. H5CY5BBXY_2_25_2.fastq.gz</li>
</ul>
<p>Generate an empty file like this:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  BARCODES <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, </span>
<span>  INDIVIDUALS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>,</span>
<span>  FORWARD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>,</span>
<span>  REVERSE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_tsv</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.</span>, path <span class="op">=</span> <span class="st">"02_project_info/project.info.turtle.tsv"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="faq">FAQ<a class="anchor" aria-label="anchor" href="#faq"></a>
</h4>
<div class="section level5">
<h5 id="multiple-sequencing-lanes-or-chips">Multiple sequencing lanes or chips ?<a class="anchor" aria-label="anchor" href="#multiple-sequencing-lanes-or-chips"></a>
</h5>
<p>If you have several Illumina lanes or Ion Torrent chips, just write
them down along the sample and barcodes associated with it. On the same
file.</p>
</div>
<div class="section level5">
<h5 id="technical-replicates">Technical replicates ?<a class="anchor" aria-label="anchor" href="#technical-replicates"></a>
</h5>
<p>2 options:</p>
<ul>
<li>use separate name for it or best,</li>
<li>let stackr manage it my using the same name (yes… yes…).</li>
</ul>
<p>stackr will see that you have, for different barcodes/lanes combo,
ids that have the exact same name. What will happen ?</p>
<ul>
<li>things will be kept separate and <code>-1</code> and <code>-2</code>
will be added to the sample name.</li>
<li>after <code>run_process_radtags</code>, the replicates will be
combined and the new file will have <code>-R</code> at the end. The
individual replicates are kept separate as well.</li>
</ul>
<p><strong>example:</strong></p>
<p>You have a sample, <code>STU-BUR-2019-F-0001</code>, on 2 separate
wells on the the plate or even lanes, with different barcode. At the end
of <code>run_process_radtags</code> you’ll have:</p>
<ul>
<li><code>STU-BUR-2019-F-0001-1</code></li>
<li><code>STU-BUR-2019-F-0001-2</code></li>
<li>
<code>STU-BUR-2019-F-0001-R</code> (for the combined
replicates)</li>
</ul>
</div>
<div class="section level5">
<h5 id="no-lanes-just-individuals-fq-files">No lanes just individual’s fq files ?<a class="anchor" aria-label="anchor" href="#no-lanes-just-individuals-fq-files"></a>
</h5>
<p>Example: you used DArT sequencing services. You have individual fq
files, the barcodes are still in there along some Illumina adapters.</p>
<p>Use the <code>LANES</code> column for the individual FQ files that
requires to go through stacks process_radtags.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="transfer-sequencing-files">Transfer sequencing files<a class="anchor" aria-label="anchor" href="#transfer-sequencing-files"></a>
</h3>
<p>Transfer the sequencing files inside the folder:
<code>03_sequencing_lanes</code>. Contrary to the pipeline I used to
contribute and follow <a href="https://github.com/enormandeau/stacks_workflow#copy-your-raw-files" class="external-link">link</a>
<strong>Don’t modify the files</strong>: stackr, like stacks, handles
the different types.</p>
</div>
<div class="section level3">
<h3 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h3>
<p>I highly recommend running the sequencing files (entire lane or chip)
through a software dedicated for QC. If you have bad sequencing, no need
to go further in the pipeline…</p>
<ul>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>.</li>
<li>
<a href="http://www.sthda.com/english/wiki/fastqcr-an-r-package-facilitating-quality-controls-of-sequencing-data-for-large-numbers-of-samples" class="external-link">fastqcr</a>
is the R alternative.</li>
</ul>
</div>
<div class="section level3">
<h3 id="run_process_radtags">run_process_radtags<a class="anchor" aria-label="anchor" href="#run_process_radtags"></a>
</h3>
<p>This is the demultiplexing step in stacks, the module is called <a href="http://catchenlab.life.illinois.edu/stacks/comp/process_radtags.php" class="external-link">process_radtags</a>.</p>
<p>Function documentation: <code><a href="../reference/run_process_radtags.html">?stackr::run_process_radtags</a></code>
(read it twice).</p>
<p><strong>Note:</strong></p>
<p>I really like <a href="https://cutadapt.readthedocs.org" class="external-link">cutadapt</a>, it’s my working
horse for reads operations. However, <a href="https://github.com/enormandeau/stacks_workflow#running-cutadapt" class="external-link">I
no longer recommend</a>, using it, unless you have very very problematic
fastq files or need to do very specific operations on your reads. Stay
with stacks, <a href="http://catchenlab.life.illinois.edu/stacks/comp/process_radtags.php" class="external-link">process_radtags</a>
will handle your fastq files just fine.</p>
<p>Difference with stacks <code>process_radtags</code> ?</p>
<ul>
<li>You have one function that will run all you lanes/chips.</li>
<li>Technical replicates will be managed and combined.</li>
<li>A summary of the process will be generated.</li>
</ul>
<p>The function will generate, inside <code>08_stacks_results</code>,
files like these:</p>
<ul>
<li><code>process_radtags_LANE_01_barcodes_stats.tsv</code></li>
<li><code>process_radtags_LANE_01_stats.tsv</code></li>
</ul>
<p>These files have the statistics per barcodes and per lanes. The logs
are found inside the folder: <code>09_log_files</code>.</p>
<p>The project info file will be populated with these columns/info:</p>
<ul>
<li>
<code>TOTAL</code>: the number of total reads.</li>
<li>
<code>NO_RADTAG</code>: the number of reads without radtags.</li>
<li>
<code>LOW_QUALITY</code>: the number of reads discarded because of
low quality.</li>
<li>
<code>RETAINED</code>: The number of reads retained after the
filters used inside process_radtags.</li>
<li>
<code>SQL_ID</code>: A unique integer that will be used inside the
next step: <code>run_ustacks</code>
</li>
<li>
<code>FQ_FILES_F</code>: the individual forward fq files</li>
<li>
<code>FQ_FILES_R</code>: the individual reverse fq files</li>
</ul>
<div class="section level4">
<h4 id="reads-length">reads length<a class="anchor" aria-label="anchor" href="#reads-length"></a>
</h4>
<p>If you are using Illumina sequencing, chances are that you have 100
pb reads (SE RADseq). With Ion Torrent, the size will be highly variable
and you have to determine the optimal size to cut your reads to the same
length (<a href="http://catchenlab.life.illinois.edu/stacks/manual/" class="external-link">stacks works
better with uniform read length, check stacks manual under section
<em>Sequencer Type</em></a>).</p>
<ol style="list-style-type: decimal">
<li>output of <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
</li>
</ol>
<p><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
will output the read length distribution. But it’s difficult to know how
many markers you will get depending on read size.</p>
<p>You will see the relationship between the position of the nucleotide
on the read and overall quality. This is very helpful to determine if
you want to cut all your Illumina reads to 80 or 90 pb.</p>
<p><strong>Below is a FastQC figure of a typical Ion Torrent sequencing
chip sequence length distribution:</strong>
<img src="ion_torrent_1.png"></p>
<ol start="2" style="list-style-type: decimal">
<li>Use stackr</li>
</ol>
<p>Below is a figure generated by stackr that shows the number of reads
you will get depending on the maximum size threshold selected:</p>
<p><img src="ion_torrent_2.png"></p>
<p>The function is:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">size</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/reads_length_distribution.html">reads_length_distribution</a></span><span class="op">(</span></span>
<span>  fq.file <span class="op">=</span> <span class="st">"my-fastq.fastq.gz"</span>, </span>
<span>  with.future <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  parallel.core <span class="op">=</span> <span class="fl">12L</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>You can use individual file or entire chip or lane. Function
documentation: <code><a href="../reference/reads_length_distribution.html">?stackr::reads_length_distribution</a></code></p>
<p><strong>Remember that depending of types of file (individual or
entire lane/chip:</strong></p>
<ul>
<li>Barcodes and adapters might be in this relationship observed in the
figure.</li>
<li>It’s not uncommon to see increasing adapters present with increasing
read size.</li>
<li>Repetitive sequences, unique reads and so on might impact the
results (<a href="#noise">see noise section below</a>).</li>
</ul>
</div>
<div class="section level4">
<h4 id="se-radseq">SE RADseq<a class="anchor" aria-label="anchor" href="#se-radseq"></a>
</h4>
<p>Example with single-end, 2 enzymes RADseq//GBS data:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">step1</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/run_process_radtags.html">run_process_radtags</a></span><span class="op">(</span></span>
<span>  project.info <span class="op">=</span> <span class="st">"02_project_info/project.info.turtle.tsv"</span>,</span>
<span>  path.seq.lanes <span class="op">=</span> <span class="st">"03_sequencing_lanes"</span>,</span>
<span>  q <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  renz.1 <span class="op">=</span> <span class="st">"pstI"</span>, </span>
<span>  renz.2 <span class="op">=</span> <span class="st">"sphI"</span>,</span>
<span>  t <span class="op">=</span> <span class="fl">90</span>, </span>
<span>  len.limit <span class="op">=</span> <span class="fl">90</span>,</span>
<span>  filter.illumina <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  retain.header <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  adapter.1 <span class="op">=</span> <span class="st">"CGAGATCGGAAGAGCGGG"</span>, </span>
<span>  adapter.mm <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fl">12</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># look for the remaining default arguments...</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="pe-radseq">PE RADseq<a class="anchor" aria-label="anchor" href="#pe-radseq"></a>
</h4>
<p>Example with paired-end 1 enzyme RADseq:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">step1</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/run_process_radtags.html">run_process_radtags</a></span><span class="op">(</span></span>
<span>  P <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  project.info <span class="op">=</span> <span class="st">'02_project_info/project.info.turtle.tsv'</span>,</span>
<span>  path.seq.lanes <span class="op">=</span> <span class="st">"03_sequencing_lanes"</span>,</span>
<span>  enzyme <span class="op">=</span> <span class="st">"sbfI"</span>,</span>
<span>  adapter.1 <span class="op">=</span> <span class="st">"CGAGATCGGAAGAGCGGG"</span>,</span>
<span>  adapter.2 <span class="op">=</span> <span class="st">"CGAGATCGGAAGAGCGGG"</span>,</span>
<span>  adapter.mm <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  t <span class="op">=</span> <span class="fl">90</span>, </span>
<span>  len.limit <span class="op">=</span> <span class="fl">90</span>,</span>
<span>  retain.header <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fl">8</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># look for the remaining default arguments...</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="qc">QC<a class="anchor" aria-label="anchor" href="#qc"></a>
</h4>
<p>I highly recommend running the individual fastq files through <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
or <a href="http://www.sthda.com/english/wiki/fastqcr-an-r-package-facilitating-quality-controls-of-sequencing-data-for-large-numbers-of-samples" class="external-link">fastqcr</a>.</p>
</div>
</div>
<div class="section level3">
<h3 id="noise">Noise reduction<a class="anchor" aria-label="anchor" href="#noise"></a>
</h3>
<p>Reducing the noise in an individual FQ file could really help
speeding up your analysis. I use it when I have more than 200 samples.
Nothing new here:</p>
<ul>
<li>Daniel Ilut explains it in his pipeline (Ilut et al. 2014) and my
function here his basically the same as his but in R…</li>
<li>Jon Puritz’s <a href="http://www.ddocent.com/UserGuide/#de-novo-assembly" class="external-link">dDocent data
reduction step</a> as a similar example.</li>
<li>Praveen’s pipeline <a href="https://github.com/beiko-lab/RADProc" class="external-link">RADProc</a> (Ravindran et
al. 2019) use something similar.</li>
<li>Metagenomic analysis pipeline all have some step of denoising the
data.</li>
</ul>
<div class="section level4">
<h4 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h4>
<p>I prefer Daniel Ilut way of visualizing the information.</p>
<p>The function <code><a href="../reference/read_depth_plot.html">stackr::read_depth_plot</a></code> requires 1 FQ file
(after stacks <em>process_radtags</em> or
<code><a href="../reference/run_process_radtags.html">stackr::run_process_radtags</a></code>) and <strong>is very fast to
run</strong>. It will highlight read coverage groups of distinct reads
within a sample. When I say distinct reads, is reads that are exactly
the same, no mismatches. The distinct reads are putative allele.</p>
<p><img src="read_depth_coverage_groups.png"></p>
<p><strong>4 read coverage groups are shown:</strong></p>
<ul>
<li><p><strong>distinct reads with low coverage (in red): </strong>
these reads are likely sequencing errors or uninformative polymorphisms
(shared only by a few samples).</p></li>
<li>
<p><strong>disting reads for a target coverage (in green):
</strong></p>
<ul>
<li>Usually represent around 80% of the reads in the FQ file.</li>
<li>It’s a <em>safe coverage range</em> to start exploring your data
(open for discussion).</li>
<li>Lower threshold (default = 7): you could go as low as maybe 3 ? You
can’t escape it, it’s your tolerance to call heterozygote <strong>a true
heterozygote</strong>. You want a minimum coverage for both the
reference and the alternative allele. Yes, you can use population
information to lower this threshold or use some fancy <em>bayesian</em>
algorithm.</li>
<li>Higher threshold: is a lot more open for discussion, here it’s the
lower limit of another group (the orange, see below for description).
Minus 1 bp.</li>
</ul>
</li>
<li><p><strong>distinct reads with high coverage &gt; 1 read depth (in
yellow): </strong> those are legitimate alleles with high
coverage.</p></li>
<li><p><strong>distinct and unique reads with high coverage (in orange):
</strong> this is where you’ll find paralogous sequences, transposable
elements and other high copy number variants.</p></li>
</ul>
<p>To run it on your data:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">read.plot</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/read_depth_plot.html">read_depth_plot</a></span><span class="op">(</span></span>
<span>  fq.file <span class="op">=</span> <span class="st">"04_process_radtags/my-fq.fq.gz"</span>,</span>
<span>  min.coverage.fig <span class="op">=</span> <span class="fl">7L</span>, <span class="co">#default</span></span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span> <span class="co">#default</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The function will generate the plot shown above and write the figure
in the working directory, using the sample name:
<code>my-fq_hap_read_depth.png</code>.</p>
</div>
<div class="section level4">
<h4 id="faq-1">FAQ<a class="anchor" aria-label="anchor" href="#faq-1"></a>
</h4>
<p><strong>Should I run the function on all samples ?</strong></p>
<p>Please don’t. Their is no use in running this on the entire dataset.
It’s going to be fairly the same. Run it on a couple of samples. Based
on the number of reads (distribution), choose 1 sample: below average,
average and above average and compare the look of the figures
produced.</p>
</div>
<div class="section level4">
<h4 id="filtering">Filtering ?<a class="anchor" aria-label="anchor" href="#filtering"></a>
</h4>
<p>Removing the noise in FQ files, it’s usually a hot topic… Personally,
the less I do the better and it’s definitely the case for small
projects. I don’t think the noise is a problem for stacks. For large
projects, it’s something else, and certainly a huge bottleneck. When you
generate <em>de novo</em> assembly for more than 200 samples and
generate a catalog with all the noise, cleaning the data can represent
weeks of computation time that you shave off your project… really.</p>
<p>You don’t have to believe what I just wrote, test it yourself:</p>
<p><strong>1. Clean the FQ file</strong></p>
<p>Use the sample you just used in the read coverage groups, above,
clean it with <code><a href="../reference/clean_fq.html">stackr::clean_fq</a></code>. It’s very fast to run:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clean.fq.name</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/clean_fq.html">clean_fq</a></span><span class="op">(</span></span>
<span>  fq.file <span class="op">=</span> <span class="st">"04_process_radtags/my-fq.fq.gz"</span>, </span>
<span>  min.coverage.threshold <span class="op">=</span> <span class="fl">2</span>, </span>
<span>  remove.unique.reads <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co">#default</span></span>
<span>  write.blacklist.fasta <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  write.blacklist <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co">#default</span></span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>, <span class="co">#default</span></span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span> <span class="co">#default</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>This will generate:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my</span><span class="op">-</span><span class="va">fq_cleaned.fq.gz</span> <span class="co"># the cleaned fq file</span></span>
<span><span class="va">my</span><span class="op">-</span><span class="va">fq_blacklisted_reads.fasta.gz</span> <span class="co"># a fasta file with blacklisted sequences...</span></span></code></pre></div>
<p><strong>2. <em>de novo</em> assembly</strong></p>
<p>Run stacks <em>ustacks</em> or even better, stay in stackr and do the
next step below: <em>the mismatch threshold series</em> (using
<code><a href="../reference/run_ustacks.html">stackr::run_ustacks</a></code>). For this, use the raw and clean FQ
files and compare the output…</p>
<p>In all my projects, having less reads (the cleaned FQ file)
means:</p>
<ul>
<li>faster <em>de novo</em> assemblies (I’ve seen up to 3 times
faster).</li>
<li>less loci generated, but of better qualities:
<ul>
<li>higher coverage in all the <em>de novo</em> assembly steps of
stacks</li>
<li>less repetitive stacks blacklisted</li>
<li>lower number of locus with assembly artifacts (lors of SNPs/locus,
etc)</li>
<li>very similar heterozygosity and homozygosity proportions</li>
</ul>
</li>
<li>the mismathches plots are almost always identical: showing me that I
can use the same Mismatch threshold (the <code>-M</code> in stacks
<em>ustacks</em>) for the cleaned or the raw FQ files.</li>
<li>faster catalog generation time, the stacks <em>cstacks</em> or
<code><a href="../reference/run_cstacks.html">stackr::run_cstacks</a></code> step.</li>
<li>NO DIFFERENCES in downstream analysis (population structure, outlier
detection, Fst, assignment, closekin).</li>
</ul>
<p><strong>Why keep the blacklisted sequences (the fasta file)
?</strong></p>
<p>You could use <code><a href="https://rdrr.io/pkg/Biostrings/man/matchPattern.html" class="external-link">Biostrings::countPattern</a></code> function to look
for TE in those sequences;)</p>
<p><strong>3. Filtering the entire dataset</strong></p>
<p>If you see a big difference and are convince that filtering is the
way to go, you can easily do it with this piece of code:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fq.to.clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"04_process_radtags/"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fq.to.clean</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">fq.to.clean</span></span>
<span></span>
<span><span class="va">clean.all</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map_dfr.html" class="external-link">map_df</a></span><span class="op">(</span></span>
<span>  .x <span class="op">=</span> <span class="va">fq.to.clean</span>, </span>
<span>  .f <span class="op">=</span> <span class="fu">stackr</span><span class="fu">::</span><span class="va"><a href="../reference/clean_fq.html">clean_fq</a></span>,</span>
<span>  min.coverage.threshold <span class="op">=</span> <span class="fl">5</span>, </span>
<span>  remove.unique.reads <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co">#default</span></span>
<span>  write.blacklist.fasta <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  write.blacklist <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co">#default</span></span>
<span>  parallel.core <span class="op">=</span> <span class="fl">12</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Transfer the cleaned files or the raw ones in a separate
folder/sub-folder to keep things tidy.</p>
</div>
</div>
<div class="section level3">
<h3 id="normalization">Data normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h3>
<p>Another hot topic, up for debate…</p>
<p>After the stacks step called <em>process_radtags</em>, in stackr
<code>run_process_radtags</code>, looking at the distribution of the
number of reads per sample, you might be getting one of these
distribution:</p>
<p><img src="read_distribution.png"></p>
<p>In an ideal world, <strong>you want the uniform
distribution</strong>. That translate in good field and wet-lab
techniques:</p>
<ul>
<li>tissue sampling and preservation</li>
<li>extraction (robots ?)</li>
<li>pipetting</li>
<li>DNA quantification</li>
<li>all the other required techniques</li>
</ul>
<p>And this resulted in similar sequenced information between your
samples. You’re about to compare apples with apples. :)</p>
<p>If you have a skewed (left or right) or a normal distribution, you’ll
be comparing apples with … citrus. The samples with lower reads will
have less loci and usually less heterozygous loci. Similarly, the
samples with higher number of reads will have more loci, more
heterozygous loci.</p>
<p>What’s the likely cause for these ?</p>
<ul>
<li>student with no prior wet-lab experience doing the DNA extraction or
downstream step (e.g. library prep)</li>
<li>data preservation issues: historical DNA, fragmented DNA, etc.</li>
<li>skipping the DNA quantification or library QC steps</li>
</ul>
<p>Ok, but what’s the problem with that ?</p>
<p><strong>Potentially these problems:</strong></p>
<ul>
<li>ascertainment bias that could drive populations polymorphism
discovery bias</li>
<li>missing data patterns</li>
<li>individual heterozygosity problems or patterns</li>
<li>downstream analysis impacted by the differences between your
samples: artifactual or biological ?</li>
</ul>
<p><strong>Solutions ?</strong></p>
<div class="section level4">
<h4 id="wet-lab-normalization">1. Wet-lab normalization<a class="anchor" aria-label="anchor" href="#wet-lab-normalization"></a>
</h4>
<p>Here are the basic steps:</p>
<ul>
<li>sequence the samples</li>
<li>look at the stats we’ve discussed</li>
<li>go back in the wet-lab to prepare a new plate</li>
<li>sequence</li>
<li>pool or not the lanes/chips</li>
<li>look at the sequencing statistics</li>
<li>repeat the step above or continue with the pipeline</li>
</ul>
<p>I’m not a fan:</p>
<ul>
<li>Way to many wet-lab steps.</li>
<li>The wet-lab is <strong>THE MOST</strong> important part.</li>
<li>Everything gets exacerbated with RADseq.</li>
<li>I’m not in favor of pooling lanes or chips (more on this some
day!).</li>
<li>It’s really not like working with microsatelites.</li>
</ul>
</div>
<div class="section level4">
<h4 id="bio-informatic-normalization">2. Bio-informatic normalization<a class="anchor" aria-label="anchor" href="#bio-informatic-normalization"></a>
</h4>
<p>When doing it right from the start is impossible, I prefer to do the
normalization after de demultiplexing step… after
<code><a href="../reference/run_process_radtags.html">stackr::run_process_radtags</a></code>. The prerequisite is to have
enough sequencing material (reads) for all my samples.</p>
<p>The tool? <a href="http://thierrygosselin.github.io/stackr/reference/normalize_reads.html" class="external-link">stackr::normalize_reads</a>
function.</p>
<p>Rarefaction of fasq files by sub-sampling the reads before <em>de
novo</em> assembly or alignment is not new:</p>
<ul>
<li>
<strong>Community and biodiversity ecologists</strong> are well
aware of species accumulation curves obtained with rarefaction (Sanders,
1968). The rarefaction method guarantees the samples have the same
weight in the analysis.</li>
<li>
<strong>Metagenomics community</strong> before analyzing the alpha
diversity will deal with differing sample depth for a given diversity
statistic by using rarefaction. It means taking a random subset of a
given size of the original sample (see <a href="https://mothur.org" class="external-link">mothur</a>, Schloss et al. 2009).</li>
<li>
<strong>Molecular ecologists</strong> for some reasons, RADseq
enthusiasts seems unaware of this concept, but see (Hale et al. 2009;
Kalinowski, 2004 and Puncher et al. 2018).</li>
<li>
<strong>The Keywords:</strong> rarefaction, sub-sampling,
normalization, standardization, sample size correction.</li>
</ul>
<p><strong>The steps ?</strong></p>
<ul>
<li>the goal is to reduce variance in number of reads between
samples</li>
<li>rarefaction/normalization is conducted aon the fq file after stacks
<em>process_radtags</em> and after removing the noise.</li>
<li>normalized replicates are generated by the function by randomly
selecting the reads (without replacement) for each samples</li>
<li>you choose the number of replicates (default: 3) and you manage
those replicates like other replicates you might have (you want
replicates…).</li>
</ul>
<p>Read the <a href="http://thierrygosselin.github.io/stackr/reference/normalize_reads.html#examples" class="external-link">function
documentation and look at the example</a></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># To run this function, bioconductor's ShortRead package is necessary:</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ShortRead"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># using defaults:</span></span>
<span><span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/normalize_reads.html">normalize_reads</a></span><span class="op">(</span>path.samples <span class="op">=</span> <span class="st">"~/corals"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># customizing the function:</span></span>
<span><span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/normalize_reads.html">normalize_reads</a></span><span class="op">(</span></span>
<span>   project.info <span class="op">=</span> <span class="st">"project.info.corals.tsv"</span>,</span>
<span>   path.samples <span class="op">=</span> <span class="st">"04_process_radtags"</span>,</span>
<span>   sample.reads <span class="op">=</span> <span class="fl">2000000</span>,</span>
<span>   number.replicates <span class="op">=</span> <span class="fl">3</span>,</span>
<span>   random.seed <span class="op">=</span> <span class="fl">3</span>,</span>
<span>   parallel.core <span class="op">=</span> <span class="fl">12</span></span>
<span>   <span class="op">)</span></span></code></pre></div>
<p><strong>My observations using rarefaction:</strong></p>
<ul>
<li>the replicates have similar mismathches plots: showing me that I can
use the same Mismatch threshold (the <code>-M</code> in stacks
<em>ustacks</em>) for the rarefied samples.</li>
<li>better <em>de novo</em> assembly statistics</li>
<li>when the individual or population heterozygosity <strong>is a real
biological signal</strong>, rarefaction <strong>doesn’t impact the
downstream analysis</strong>.</li>
<li>when the heterozygosity was driven by the sequencing bias,
bioinformatic rarefaction took care of the problem.</li>
</ul>
<div class="section level5">
<h5 id="faq-2">FAQ<a class="anchor" aria-label="anchor" href="#faq-2"></a>
</h5>
<p><strong>It doesn’t bother you to throw away all the valuable
information that samples with high read numbers have ?</strong></p>
<p>Absolutely not!</p>
<ul>
<li>
<strong>most of the information is already lost</strong>: that is if
you used stacks to generate a catalog and then filtered the data to keep
markers in x individuals and/or in all sampling sites/populations. At
this step it’s either you loose markers or get more missing data…</li>
<li>
<strong>It’s not a loss of information if the information cannot be
confirmed with other samples:</strong> it’s a statistic problem and
ecologists are well aware of this.</li>
<li>
<strong>The loss of reads is intrinsic to RADseq: </strong> Don’t
believe me, inspect stacks, stackr or any RADseq pipeline log files…
Rarefaction is just another step, molecular ecologists are not used to
it.</li>
<li>check the total read depth of 1 sample after all the bioinformatic
steps. Compare the numbers with the actual read number from the start of
the pipeline. If you have &gt; 30% you are lucky!</li>
</ul>
</div>
</div>
</div>
<div class="section level3">
<h3 id="run_ustacks">run_ustacks<a class="anchor" aria-label="anchor" href="#run_ustacks"></a>
</h3>
<p>This is the <em>de novo</em> assembly step in stacks, the module is
called <a href="http://catchenlab.life.illinois.edu/stacks/comp/ustacks.php" class="external-link">ustacks</a>.</p>
<p>Function documentation: <code><a href="../reference/run_ustacks.html">?stackr::run_ustacks</a></code> (read it
twice).</p>
<p>Difference with stacks <code>ustacks</code> ?</p>
<ul>
<li>SQL IDs or the <code>-i</code> argument, the unique integer ID for
this sample is managed automatically.</li>
<li>computer/cluster/server crash ? Just re-run the function…</li>
<li>The mismatch testing can be done inside
<code><a href="../reference/run_ustacks.html">stackr::run_ustacks</a></code> automatically ;)</li>
<li>A summary is generated and can be run separately for those who ran
the function inside the terminal with stacks. Documentation:
<code><a href="../reference/summary_ustacks.html">?summary_ustacks</a></code>.</li>
</ul>
<p>If you don’t know the optimal number of mismatches between your
individual reads this part is the most important steps of <em>de
novo</em> assembly.</p>
<div class="section level4">
<h4 id="mismatches">Mismatches<a class="anchor" aria-label="anchor" href="#mismatches"></a>
</h4>
<p>When using paired-end data, use forward reads for the analysis.</p>
<p>Select one sample to do the test. For this, you could use a sample
with average file size or use the stats produced directly by the output
of <code>run_process_radtags</code> and available inside the updated
project info file saved by the function.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">info.stats</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"02_project_info/project.info.20191121@1606.tsv"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">info.stats</span>, <span class="va">REPLICATES</span> <span class="op">==</span> <span class="st">"1"</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">info.stats</span><span class="op">$</span><span class="va">RETAINED</span><span class="op">)</span></span>
<span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">info.stats</span><span class="op">$</span><span class="va">RETAINED</span><span class="op">)</span></span>
<span></span>
<span><span class="va">read.min</span> <span class="op">&lt;-</span> <span class="va">m</span> <span class="op">-</span> <span class="va">s</span></span>
<span><span class="va">read.max</span> <span class="op">&lt;-</span> <span class="va">m</span> <span class="op">+</span> <span class="va">s</span></span>
<span></span>
<span><span class="va">sample.select</span> <span class="op">&lt;-</span> <span class="va">info.stats</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">RETAINED</span> <span class="op">&gt;=</span> <span class="va">read.min</span> <span class="op">&amp;</span> <span class="va">RETAINED</span> <span class="op">&lt;=</span> <span class="va">read.max</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/sample_n.html" class="external-link">sample_n</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">sample.select</span><span class="op">$</span><span class="va">FQ_FILES_F</span></span>
<span><span class="co"># [1] "BET-LAM-60044-1682200-1.1.fq.gz"</span></span></code></pre></div>
<p>Make a copy the file inside the folder:
<code>05_clustering_mismatches</code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span></span>
<span>  from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"04_process_radtags"</span>, <span class="va">sample.select</span><span class="op">$</span><span class="va">FQ_FILES_F</span><span class="op">)</span>, </span>
<span>  to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"05_clustering_mismatches/"</span>, <span class="va">sample.select</span><span class="op">$</span><span class="va">FQ_FILES_F</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Run the mismatch clustering series:</strong></p>
<ul>
<li>with files around 2M reads</li>
<li>testing 1 to 5 mismatches is very fast to run</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mismatch</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/run_ustacks.html">run_ustacks</a></span><span class="op">(</span></span>
<span>  mismatch.testing <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  f <span class="op">=</span> <span class="st">"05_clustering_mismatches"</span>,</span>
<span>  M <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>  model.type <span class="op">=</span> <span class="st">"bounded"</span><span class="op">)</span></span></code></pre></div>
<p>This will generate 5 folders inside
<code>05_clustering_mismatches</code> that will have the mismatches
number and associated ustacks output files in it.</p>
<p>A figure similar to this one:</p>
<p><img src="mismatches.plot.png"></p>
<p>To understand this figure, I suggest you read these 2 excellent
papers:</p>
<p>Ilut, D., Nydam, M., Hare, M. (2014). Defining Loci in
Restriction-Based Reduced Representation Genomic Data from Non model
Species: Sources of Bias and Diagnostics for Optimal Clustering BioMed
Research International 2014(2), 1 9. <a href="https://dx.doi.org/10.1155/2014/675158" class="external-link uri">https://dx.doi.org/10.1155/2014/675158</a></p>
<p>Harvey, M., Judy, C., Seeholzer, G., Maley, J. (2015). Similarity
thresholds used in short read assembly reduce the comparability of
population histories across species PeerJ 3(), e895. <a href="https://dx.doi.org/10.7717/peerj.895" class="external-link uri">https://dx.doi.org/10.7717/peerj.895</a></p>
<p>The function also produce a tibble
<code>mismatches.summary.tsv</code> with several columns and lines
summarizing ustacks output logs for the different mismatches.</p>
<p><img src="mismatches.summary.1.png"></p>
<p>This is the data used to make the figure above.</p>
<p><img src="mismatches.summary.2.png"></p>
</div>
<div class="section level4">
<h4 id="de-novo-assembly">de novo assembly<a class="anchor" aria-label="anchor" href="#de-novo-assembly"></a>
</h4>
<p>Now that you have decided the mismatch threshold, corresponding to
the <code>M</code> argument inside <code><a href="../reference/run_ustacks.html">stackr::run_ustacks</a></code>, run
the function on all samples.</p>
<p>Depending on project size, you might want to look at the alternative
of ustacks and cstacks: <a href="#alternative">run_radproc</a>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">u</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/run_ustacks.html">run_ustacks</a></span><span class="op">(</span></span>
<span>  project.info <span class="op">=</span> <span class="st">"02_project_info/project.info.20191121@1606.tsv"</span>,</span>
<span>  M <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  model.type <span class="op">=</span> <span class="st">"bounded"</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fl">12</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># careful with the defaults...</span></span></code></pre></div>
<p>When using paired-end data, the function uses the forward reads for
the assembly. The reverse reads are used later in the pipeline.</p>
<p><strong>Output: </strong></p>
<p>The folder <code>06_ustacks_2_gstacks</code> will be populated with 3
files per samples ending with:</p>
<ul>
<li><code>alleles.tsv.gz</code></li>
<li><code>tags.tsv.gz</code></li>
<li><code>snps.tsv.gz</code></li>
</ul>
<p>These files are <a href="http://catchenlab.life.illinois.edu/stacks/manual/#files" class="external-link">documented
here</a></p>
<p>The folder <code>08_stacks_results</code> will have a summary file:
<code>summary_ustacks_date@time.tsv</code>.</p>
<p>The folder <code>09_log_files</code> as the associated individual log
files.</p>
</div>
</div>
<div class="section level3">
<h3 id="run_cstacks">run_cstacks<a class="anchor" aria-label="anchor" href="#run_cstacks"></a>
</h3>
<p>This is the catalog or pseudo reference genome for the <em>de
novo</em> part of stacks, the module is called <a href="http://catchenlab.life.illinois.edu/stacks/comp/cstacks.php" class="external-link">cstacks</a>.
It’s probably the longest part of the pipeline for large sample-size
projects. It’s also the part where a crashing computer or cluster as the
biggest drawbacks and test your bioinformatic patience.</p>
<p>Depending on project size, you might want to look at the alternative
of ustacks and cstacks: <a href="#alternative">run_radproc</a>.</p>
<p>Function documentation: <code><a href="../reference/run_cstacks.html">?stackr::run_cstacks</a></code> (read it
twice).</p>
<p>Difference with stacks <code>cstacks</code> ?</p>
<ul>
<li>Generating the catalog is split in x number of samples to make it
more robust.</li>
</ul>
<p><strong>How?</strong></p>
<ul>
<li>The population map that shows which samples to use to build the
catalog is split in 20 individuals (by default).</li>
<li>The function generates a catalog after the first 20 samples and
writes it down.</li>
<li>
<code>cstacks</code> starts again with the previous catalog and
build a new one with new samples next on the list.</li>
<li>It’s incremental</li>
<li>Saves you a lot of time if e.g. your computer crash.</li>
<li>You can easily start with the catalog that was last written in the
folder.</li>
</ul>
<p>For now, I highly discourage subsampling to build the catalog =
please use all the samples in the project.</p>
<p>Start by generating a population map:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pop.map.catalog</span> <span class="op">&lt;-</span> <span class="va">u</span><span class="op">$</span><span class="va">project.info</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">INDIVIDUALS_REP</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>POP_ID <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_tsv</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">.</span>,</span>
<span>    path <span class="op">=</span> <span class="st">"02_project_info/population.map.turtle.catalog.tsv"</span>,</span>
<span>    col_names <span class="op">=</span> <span class="cn">FALSE</span><span class="co"># stacks doesn't want column header</span></span>
<span>  <span class="op">)</span> </span></code></pre></div>
<p><strong>Build the catalog:</strong></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">c</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/run_cstacks.html">run_cstacks</a></span> <span class="op">(</span></span>
<span>  P <span class="op">=</span> <span class="st">"06_ustacks_2_gstacks"</span>,</span>
<span>  M <span class="op">=</span> <span class="st">"02_project_info/population.map.turtle.catalog.tsv"</span>,</span>
<span>  n <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fl">32</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Who’s in the catalog?</strong></p>
<p>Use the function <a href="http://thierrygosselin.github.io/stackr/reference/extract_catalog_sql_ids.html" class="external-link">stackr::extract_catalog_sql_ids</a>
to verify the SQL IDs of the samples used for the catalog
construction.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu">extract_catalog_sql_id</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"06_ustacks_2_gstacks/catalog.tags.tsv.gz"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="alternative">ustacks/cstacks alternatives ?<a class="anchor" aria-label="anchor" href="#alternative"></a>
</h3>
<p>You have tried the noise reduction <a href="#noise">step</a> and/or
the normalization <a href="#normalization">step</a> and it’s still way
to long ?</p>
<p>The <a href="https://github.com/beiko-lab/RADProc" class="external-link">RADProc</a>
approach developped by Praveen Nadukkalam Ravindran (Ravindran et
al. 2019) is a viable alternative and it’s compatible with stacks. You
replace the <em>de novo</em> and catalog steps by
<strong>RADProc</strong>. For a project I did, it’s a question of 15
days to run stacks <strong>ustacks/cstacks</strong> or 1,5 days to run
<strong>RADProc</strong> + the rest of stacks pipeline… no kidding!</p>
<p><strong>Limitations</strong></p>
<ul>
<li>
<em>de novo</em> only</li>
<li>Works only with SE RADseq data, no paired-end.</li>
<li>No gapped alignment (for INDELS):
<ul>
<li>some sequencing technologies are more prone to INDEL…</li>
<li>verify by generating a test set with both approach and
comparing.</li>
</ul>
</li>
<li>No control on SQL IDs (the id use for the catalog).</li>
<li>You cannot add sample to a pre-existing catalog: you have to start
from the start with <em>de novo</em> assemblies of all samples you want
in the catalog…</li>
</ul>
<p>The current version works with stacks 2 (sstacks and after).</p>
<p>macOS installation instructions:</p>
<ul>
<li><strong>installing RADProc requires <a href="https://www.gnu.org/software/m4/m4.html" class="external-link">m4</a></strong></li>
</ul>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="bu">cd</span> Downloads</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-L</span> ftp://ftp.gnu.org/gnu/m4/m4-1.4.19.tar.gz <span class="kw">|</span> <span class="fu">tar</span> xf <span class="at">-</span></span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a><span class="bu">cd</span> m4<span class="pp">*</span></span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a><span class="ex">./configure</span> ac_cv_type_struct_sched_param=yes <span class="kw">&amp;&amp;</span> <span class="fu">make</span> <span class="at">-j28</span> <span class="kw">&amp;&amp;</span> <span class="fu">sudo</span> make install</span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a><span class="fu">sudo</span> rm <span class="at">-R</span> m<span class="pp">*</span></span></code></pre></div>
<ul>
<li>Download <a href="https://github.com/beiko-lab/RADProc" class="external-link">RADProc</a>
</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="fu">sudo</span> chmod <span class="at">-R</span> 777 RADProc-master</span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a><span class="bu">cd</span> RADProc-master/Install/</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a><span class="ex">autoreconf</span> <span class="at">--install</span></span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a><span class="fu">sh</span> ./configure</span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a><span class="fu">make</span> <span class="kw">&amp;&amp;</span> <span class="fu">sudo</span> make install</span></code></pre></div>
<p>Read the documentation…</p>
<p><strong>To run in command line</strong></p>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="ex">RADProc</span> <span class="at">-t</span> <span class="st">"gzfastq"</span> <span class="at">-f</span> <span class="st">"04_process_radtags"</span> <span class="at">-o</span> <span class="st">"06_ustacks_2_gstacks"</span> <span class="at">-M</span> 4 <span class="at">-m</span> 3 <span class="at">-n</span> 4 <span class="at">-p</span> 8 <span class="at">-x</span> 3 <span class="at">-S</span> 80 <span class="at">-D</span> 7 <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span> <span class="kw">|</span> <span class="fu">tee</span> radproc_date@time.log</span></code></pre></div>
<p><strong>To run within stackr</strong></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">r</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/run_radproc.html">run_radproc</a></span><span class="op">(</span>M <span class="op">=</span> <span class="fl">3</span>, m <span class="op">=</span> <span class="fl">4</span>, n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p>Function documentation: <code><a href="../reference/run_radproc.html">?stackr::run_radproc</a></code> (read it
twice).</p>
<p>The output of that command:</p>
<ul>
<li>3 catalog files: <code>catalog.tags.tsv</code>,
<code>catalog.snps.tsv</code>, <code>catalog.alleles.tsv</code>.</li>
<li>3 files per individual: <code>tags.tsv</code>,
<code>snps.tsv</code>, <code>alleles.tsv</code>.</li>
<li>a file with the summary statistics.</li>
<li>a log file.</li>
</ul>
</div>
<div class="section level3">
<h3 id="run_sstacks">run_sstacks<a class="anchor" aria-label="anchor" href="#run_sstacks"></a>
</h3>
<p>That’s the <a href="http://catchenlab.life.illinois.edu/stacks/comp/sstacks.php" class="external-link">sstacks</a>
part in stacks. It’s where the output from ustacks (the individual
files) are matched against the catalog.</p>
<p>Function documentation: <code><a href="../reference/run_sstacks.html">?stackr::run_sstacks</a></code> (read it
twice).</p>
<p>Difference with stacks <code>sstacks</code> ?</p>
<ul>
<li>Not much.</li>
<li>But I like the summary it does automatically at the end. It allows
to see potential problem in the data.</li>
<li>Computer or server problem during the sstacks ? Just launch again
the same command, the function will start again, but only with the
unmatched samples!</li>
</ul>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/run_sstacks.html">run_sstacks</a></span><span class="op">(</span>parallel.core <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run_tsv2bam">run_tsv2bam<a class="anchor" aria-label="anchor" href="#run_tsv2bam"></a>
</h3>
<p>That’s the <a href="http://catchenlab.life.illinois.edu/stacks/comp/tsv2bam.php" class="external-link">tsv2bam</a>
part in stacks. The tsv2bam program will transpose data so that it is
oriented by locus, instead of by sample. This allows downstream
improvements (e.g. Bayesian SNP calling).</p>
<p>If paired-end reads are available, the <code>tsv2bam</code> program
will pull in the set of paired-end reads that are associated with each
single-end locus that was assembled <em>de novo</em>.</p>
<p>Difference with stacks <code>tsv2bam</code> ?</p>
<ul>
<li>after running this function, there’s no need to run
<code>SAMtools</code>. It’s all integrated inside
<code>run_tsv2bam</code>.</li>
<li>The function uses <code>SAMtools</code> to merge <strong>in
parallel</strong> the BAM sample files into a unique BAM catalog
file..</li>
</ul>
<p>Function documentation: <code><a href="../reference/run_tsv2bam.html">?stackr::run_tsv2bam</a></code> (read it
twice).</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># example with paired-end data:</span></span>
<span><span class="va">bam.sum</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/run_tsv2bam.html">run_tsv2bam</a></span><span class="op">(</span></span>
<span>  M <span class="op">=</span> <span class="st">"02_project_info/population.map.turtle.catalog.tsv"</span>,</span>
<span>  R <span class="op">=</span> <span class="st">"04_process_radtags/"</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fl">12</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run_gstacks">run_gstacks<a class="anchor" aria-label="anchor" href="#run_gstacks"></a>
</h3>
<p>For some reasons, if you used paired-end data, this step won’t work
for now on macOS. It’s not a stackr problem, it’s a stacks problem.
Please use linux … If you don’t have access to a linux computer, try
using the cloud for this step <a href="https://thierrygosselin.github.io/RADseq_cloud_tutorial/radseq_cloud_tutorial.html" class="external-link">documentation</a>.</p>
<p>Function documentation: <code><a href="../reference/run_gstacks.html">?stackr::run_gstacks</a></code> (read it
twice).</p>
<div class="section level4">
<h4 id="single-end-1">single-end<a class="anchor" aria-label="anchor" href="#single-end-1"></a>
</h4>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/run_gstacks.html">run_gstacks</a></span><span class="op">(</span></span>
<span>  M <span class="op">=</span> <span class="st">"02_project_info/population.map.turtle.catalog.tsv"</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fl">12</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="paired-end-1">paired-end<a class="anchor" aria-label="anchor" href="#paired-end-1"></a>
</h4>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/run_gstacks.html">run_gstacks</a></span><span class="op">(</span></span>
<span>  M <span class="op">=</span> <span class="st">"02_project_info/population.map.turtle.catalog.tsv"</span>,</span>
<span>  rm.pcr.duplicates <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rm.unpaired.reads <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  unpaired <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ignore.pe.reads <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fl">12</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="run_populations">run_populations<a class="anchor" aria-label="anchor" href="#run_populations"></a>
</h3>
<p>For this step, I like to keep it simple and get a VCF file. Filters
are best handled in my package <a href="https://thierrygosselin.github.io/radiator/index.html" class="external-link">radiator</a>.</p>
<p>Function documentation: <code><a href="../reference/run_populations.html">?stackr::run_populations</a></code> (read
it twice).</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu"><a href="../reference/run_populations.html">run_populations</a></span><span class="op">(</span></span>
<span>  M <span class="op">=</span> <span class="st">"02_project_info/population.map.turtle.catalog.tsv"</span>, </span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  r <span class="op">=</span> <span class="fl">0.1</span>, </span>
<span>  R <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  p <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  parallele.core <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filtering-1">Filtering<a class="anchor" aria-label="anchor" href="#filtering-1"></a>
</h3>
<p>Tool: my package <a href="https://thierrygosselin.github.io/radiator/" class="external-link">radiator</a></p>
</div>
<div class="section level3">
<h3 id="missing-data">Missing data<a class="anchor" aria-label="anchor" href="#missing-data"></a>
</h3>
<p>Tools for visualization and imputations: <a href="https://thierrygosselin.github.io/grur/" class="external-link">grur</a></p>
</div>
<div class="section level3">
<h3 id="population-assignment">Population assignment<a class="anchor" aria-label="anchor" href="#population-assignment"></a>
</h3>
<p>Tool: <a href="http://thierrygosselin.github.io/assigner/" class="external-link">assigner</a></p>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Ilut, D., Nydam, M., Hare, M. (2014). Defining Loci in
Restriction-Based Reduced Representation Genomic Data from Non model
Species: Sources of Bias and Diagnostics for Optimal Clustering BioMed
Research International 2014(2), 1 9. <a href="https://dx.doi.org/10.1155/2014/675158" class="external-link uri">https://dx.doi.org/10.1155/2014/675158</a></p>
<p>Rochette, N., Rivera‐Colón, A., Catchen, J. (2019). Stacks 2:
Analytical methods for paired‐end sequencing improve RADseq‐based
population genomics. Molecular Ecology <a href="https://dx.doi.org/10.1111/mec.15253" class="external-link uri">https://dx.doi.org/10.1111/mec.15253</a></p>
<p>Harvey, M., Judy, C., Seeholzer, G., Maley, J. (2015). Similarity
thresholds used in short read assembly reduce the comparability of
population histories across species PeerJ 3(), e895. <a href="https://dx.doi.org/10.7717/peerj.895" class="external-link uri">https://dx.doi.org/10.7717/peerj.895</a></p>
<p>Ravindran, P., Bentzen, P., Bradbury, I., Beiko, R. (2019). RADProc:
A computationally efficient de novo locus assembler for population
studies using RADseq data Molecular Ecology Resources 19(1), 272-282. <a href="https://dx.doi.org/10.1111/1755-0998.12954" class="external-link uri">https://dx.doi.org/10.1111/1755-0998.12954</a></p>
<p>Schloss, P.D., et al., Introducing mothur: Open-source,
platform-independent, community-supported software for describing and
comparing microbial communities. Appl Environ Microbiol, 2009.
75(23):7537-41.</p>
<p>Sanders HL (1968) Marine benthic diversity: A comparative study. The
American Naturalist 102: pp. 243–282</p>
<p>Hale et al. (2009) on the relative merits of normalization and
rarefaction in gene discovery in sturgeons. BMC genomics, 10, 203.</p>
<p>Kalinowski (2004) Counting Alleles with Rarefaction. Conservation
Genetics, 5, 539–543.</p>
<p>Puncher et al. (2018) Molecular Ecology Resources, 44, 678.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
