<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Rarefaction of reads samples — normalize_reads • stackr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Rarefaction of reads samples — normalize_reads"><meta property="og:description" content="Rarefaction of fasq files by sub-sampling the reads
before de novo assembly or alignment.
The normalization/standardization/sample size correction step allows to check
if some statistics are increasing with read numbers (e.g. heterozygous markers).
It's a very easy way to disentangle artifact from biological signal caused by
varying read numbers across samples."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stackr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/stackr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/stackr" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Rarefaction of reads samples</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/stackr/blob/HEAD/R/normalize_reads.R" class="external-link"><code>R/normalize_reads.R</code></a></small>
    <div class="hidden name"><code>normalize_reads.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Rarefaction of fasq files by sub-sampling the reads
before de novo assembly or alignment.
The normalization/standardization/sample size correction step allows to check
if some statistics are increasing with read numbers (e.g. heterozygous markers).
It's a very easy way to disentangle artifact from biological signal caused by
varying read numbers across samples.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">normalize_reads</span><span class="op">(</span></span>
<span>  project.info <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">fq.files</span>,</span>
<span>  sample.reads <span class="op">=</span> <span class="fl">1e+06</span>,</span>
<span>  number.replicates <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  random.seed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-project-info">project.info<a class="anchor" aria-label="anchor" href="#arg-project-info"></a></dt>
<dd><p>(character, path, optional) When using the stackr pipeline,
a project info file is created. This file provides all the info and stats
generated by stacks and stackr.
The project info file will be updated with the new samples.
The project info filename will be appended <code>_normalize</code>.
The file should end with <code>.tsv</code>.
If no <code>project.info</code> file is provided, the function will have to look
at the number of reads in the fastq files and this will take longer.
Default: <code>project.info = NULL</code>.</p></dd>


<dt id="arg-fq-files">fq.files<a class="anchor" aria-label="anchor" href="#arg-fq-files"></a></dt>
<dd><p>(character, path) Path of folder containing the
samples to normalize.</p></dd>


<dt id="arg-sample-reads">sample.reads<a class="anchor" aria-label="anchor" href="#arg-sample-reads"></a></dt>
<dd><p>(integer) The number of reads to pick randomly.
Default: <code>sample.reads = 1000000</code>.</p></dd>


<dt id="arg-number-replicates">number.replicates<a class="anchor" aria-label="anchor" href="#arg-number-replicates"></a></dt>
<dd><p>(interger) The number of samples to generate.
With default, if 20 samples are in the folder, 100 new samples will be generated.
Default: <code>number.replicates = 5</code>.</p></dd>


<dt id="arg-random-seed">random.seed<a class="anchor" aria-label="anchor" href="#arg-random-seed"></a></dt>
<dd><p>(integer, optional) For reproducibility, set an integer
that will be used inside function that requires randomness. With default,
a random number is generated and printed in the appropriate output.
Default: <code>random.seed = NULL</code>.</p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(optional) The number of core for parallel
programming. Each samples to normalize is sequentially treated and replicates
are generated in parallel.
By default, <code>parallel.core = parallel::detectCores() - 1</code>.
This number is adjusted automatically to the number of replicates.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>fastq files with "-1", "-2", "..." appended to the original name.
If a project info file was provided, the new replicate samples info is integrated
to the file. The modified project info file will have <code>_normalize</code> appended
to the original filename.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://thierrygosselin.github.io/stackr/" class="external-link">stackr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># To run this function, bioconductor \code{ShortRead} package is necessary:</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"http://bioconductor.org/biocLite.R"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">biocLite</span><span class="op">(</span><span class="st">"ShortRead"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Using OpenMP threads</span></span></span>
<span class="r-in"><span><span class="va">nthreads</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/CallExternal.html" class="external-link">.Call</a></span><span class="op">(</span><span class="fu">ShortRead</span><span class="fu">:::</span><span class="va">.set_omp_threads</span>, <span class="fl">1L</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/on.exit.html" class="external-link">on.exit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/CallExternal.html" class="external-link">.Call</a></span><span class="op">(</span><span class="fu">ShortRead</span><span class="fu">:::</span><span class="va">.set_omp_threads</span>, <span class="va">nthreads</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># using defaults:</span></span></span>
<span class="r-in"><span><span class="fu">stackr</span><span class="fu">::</span><span class="fu">normalize_reads</span><span class="op">(</span>fq.files <span class="op">=</span> <span class="st">"~/corals"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># customizing the function:</span></span></span>
<span class="r-in"><span><span class="fu">stackr</span><span class="fu">::</span><span class="fu">normalize_reads</span><span class="op">(</span></span></span>
<span class="r-in"><span>   project.info <span class="op">=</span> <span class="st">"project.info.corals.tsv"</span>,</span></span>
<span class="r-in"><span>   fq.files <span class="op">=</span> <span class="st">"~/corals"</span>,</span></span>
<span class="r-in"><span>   sample.reads <span class="op">=</span> <span class="fl">2000000</span>,</span></span>
<span class="r-in"><span>   number.replicates <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>   random.seed <span class="op">=</span> <span class="fl">3</span>,</span></span>
<span class="r-in"><span>   parallel.core <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># You then need to run stackr: run_ustacks, run_sstacks, run_tsv2bam, run_gstacks, run_populations</span></span></span>
<span class="r-in"><span><span class="co"># or equivalent if a reference genome.</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

