<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Run STACKS tsv2bam and merges BAM files — run_tsv2bam • stackr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Run STACKS tsv2bam and merges BAM files — run_tsv2bam"><meta property="og:description" content="Runs STACKS
tsv2bam
module and additionnally, this function will also generate a summary of
stacks tsv2bam and will merge in parallel BAM sample files into a unique
BAM catalog file using SAMtools or Sambamba.
tsv2bam
converts the data (single-end or paired-end) from being organized by sample
into being organized by locus. This allows downstream improvements
(e.g. Bayesian SNP calling)."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stackr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/stackr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/stackr" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run STACKS tsv2bam and merges BAM files</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/stackr/blob/HEAD/R/run_tsv2bam.R" class="external-link"><code>R/run_tsv2bam.R</code></a></small>
    <div class="hidden name"><code>run_tsv2bam.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Runs <a href="http://catchenlab.life.illinois.edu/stacks/" class="external-link">STACKS</a>
<a href="http://catchenlab.life.illinois.edu/stacks/stacks_v2.php" class="external-link">tsv2bam</a>
module and additionnally, this function will also generate a summary of
stacks tsv2bam and will merge in parallel BAM sample files into a unique
BAM catalog file using SAMtools or Sambamba.
<a href="http://catchenlab.life.illinois.edu/stacks/stacks_v2.php" class="external-link">tsv2bam</a>
converts the data (single-end or paired-end) from being organized by sample
into being organized by locus. This allows downstream improvements
(e.g. Bayesian SNP calling).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_tsv2bam</span><span class="op">(</span></span>
<span>  P <span class="op">=</span> <span class="st">"06_ustacks_2_gstacks"</span>,</span>
<span>  M <span class="op">=</span> <span class="st">"02_project_info/population.map.tsv2bam.tsv"</span>,</span>
<span>  R <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>  cmd.path <span class="op">=</span> <span class="st">"/usr/local/bin/samtools"</span>,</span>
<span>  h <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-p">P<a class="anchor" aria-label="anchor" href="#arg-p"></a></dt>
<dd><p>(path, character) Path to the directory containing STACKS files.
Default: <code>P = "06_ustacks_2_gstacks"</code>.
Inside the folder, you should have:</p><ul><li><p><strong>the catalog files:</strong> starting with <code>batch_</code> and ending with
  <code>.alleles.tsv.gz, .snps.tsv.gz, .tags.tsv.gz</code>;</p></li>
<li><p><strong>3 files for each samples:</strong> The sample name is the prefix for
  the files ending with:
<code>.alleles.tsv.gz, .snps.tsv.gz, .tags.tsv.gz</code>.
Those files are created in the
<a href="http://catchenlab.life.illinois.edu/stacks/comp/ustacks.php" class="external-link">ustacks</a>,
<a href="http://catchenlab.life.illinois.edu/stacks/comp/sstacks.php" class="external-link">sstacks</a> and
<a href="http://catchenlab.life.illinois.edu/stacks/comp/cxstacks.php" class="external-link">cxstacks</a>
modules.</p></li>
</ul></dd>


<dt id="arg-m">M<a class="anchor" aria-label="anchor" href="#arg-m"></a></dt>
<dd><p>(character, path) Path to a population map file.
Note that the <code>-s</code> option is not used inside <strong>stackr</strong>.
Default: <code>M = "02_project_info/population.map.tsv2bam.tsv"</code>.</p></dd>


<dt id="arg-r">R<a class="anchor" aria-label="anchor" href="#arg-r"></a></dt>
<dd><p>(path, character) Directory where to find the paired-end reads files
(in fastq/fasta/bam (gz) format).</p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(integer) Enable parallel execution with the number of threads.
Default: <code>parallel.core = parallel::detectCores() - 1</code></p></dd>


<dt id="arg-cmd-path">cmd.path<a class="anchor" aria-label="anchor" href="#arg-cmd-path"></a></dt>
<dd><p>(character, path) Provide the FULL path to SAMtools
program. See details on how to install SAMtools.
Default: <code>cmd.path = "/usr/local/bin/samtools"</code>.</p></dd>


<dt id="arg-h">h<a class="anchor" aria-label="anchor" href="#arg-h"></a></dt>
<dd><p>Display this help messsage.
Default: <code>h = FALSE</code></p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p><a href="http://catchenlab.life.illinois.edu/stacks/stacks_v2.php" class="external-link">tsv2bam</a>
returns a set of <code>.matches.bam</code> files.</p>
<p>The function <code>run_tsv2bam</code> returns a list with the number of individuals, the batch ID number,
a summary data frame and a plot containing:</p><ol><li><p>INDIVIDUALS: the sample id</p></li>
<li><p>ALL_LOCUS: the total number of locus for the individual (shown in subplot A)</p></li>
<li><p>LOCUS: the number of locus with a one-to-one relationship (shown in subplot B)
with the catalog</p></li>
<li><p>MATCH_PERCENT: the percentage of locus with a one-to-one relationship
with the catalog (shown in subplot C)</p>
<p>Addtionally, the function returns a catalog.bam file, generated
by merging all the individual BAM files in parallel.</p></li>
</ol></div>
    <div id="details">
    <h2>Details</h2>
    <p><strong>Install SAMtools</strong>
<a href="http://gbs-cloud-tutorial.readthedocs.io/en/latest/07_start_from_new_image.html#install-gbs-radseq-software-time-required-30min" class="external-link">link to detailed instructions on how to install SAMtools</a></p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Catchen JM, Amores A, Hohenlohe PA et al. (2011)
Stacks: Building and Genotyping Loci De Novo From Short-Read Sequences.
G3, 1, 171-182.</p>
<p>Catchen JM, Hohenlohe PA, Bassham S, Amores A, Cresko WA (2013)
Stacks: an analysis tool set for population genomics.
Molecular Ecology, 22, 3124-3140.</p>
<p>Li H., Handsaker B., Wysoker A., Fennell T., Ruan J., Homer N.,
Marth G., Abecasis G., Durbin R. and 1000 Genome Project Data Processing
Subgroup (2009) The Sequence alignment/map (SAM) format and SAMtools.
Bioinformatics, 25, 2078-9.</p>
<p>Li H A statistical framework for SNP calling,
mutation discovery, association mapping and population genetical parameter
estimation from sequencing data. Bioinformatics. 2011 Nov 1;27(21):2987-93.</p>
<p>A. Tarasov, A. J. Vilella, E. Cuppen, I. J. Nijman, and P. Prins.
Sambamba: fast processing of NGS alignment formats. Bioinformatics, 2015.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="http://catchenlab.life.illinois.edu/stacks/" class="external-link">STACKS</a></p>
<p><a href="http://catchenlab.life.illinois.edu/stacks/stacks_v2.php" class="external-link">stacks Version 2.0Beta6</a></p>
<p><a href="http://www.htslib.org" class="external-link">SAMtools</a></p>
<p><a href="http://lomereiter.github.io/sambamba/" class="external-link">Sambamba</a></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># The simplest form of the function:</span></span></span>
<span class="r-in"><span><span class="va">bam.sum</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu">run_tsv2bam</span><span class="op">(</span><span class="op">)</span> <span class="co"># that's it !</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

