<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Run STACKS ustacks module — run_ustacks • stackr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Run STACKS ustacks module — run_ustacks"><meta property="og:description" content="Run STACKS
ustacks
module inside R!
Inside the folder 04_process_radtags, you should have all the
individual's fastq files. Those files are created in the
process_radtags
module."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stackr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/stackr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/stackr" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run STACKS ustacks module</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/stackr/blob/HEAD/R/run_ustacks.R" class="external-link"><code>R/run_ustacks.R</code></a></small>
    <div class="hidden name"><code>run_ustacks.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run <a href="http://catchenlab.life.illinois.edu/stacks/" class="external-link">STACKS</a>
<a href="http://catchenlab.life.illinois.edu/stacks/comp/ustacks.php" class="external-link">ustacks</a>
module inside R!
Inside the folder <code>04_process_radtags</code>, you should have all the
individual's fastq files. Those files are created in the
<a href="http://catchenlab.life.illinois.edu/stacks/comp/process_radtags.php" class="external-link">process_radtags</a>
module.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_ustacks</span><span class="op">(</span></span>
<span>  mismatch.testing <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sample.list <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  project.info <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  f <span class="op">=</span> <span class="st">"04_process_radtags"</span>,</span>
<span>  o <span class="op">=</span> <span class="st">"06_ustacks_2_gstacks"</span>,</span>
<span>  m <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  M <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  N <span class="op">=</span> <span class="va">M</span> <span class="op">+</span> <span class="fl">2</span>,</span>
<span>  t <span class="op">=</span> <span class="st">"guess"</span>,</span>
<span>  R <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  H <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>  h <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  d <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  keep.high.cov <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  high.cov.thres <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  max.locus.stacks <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  k.len <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  max.gaps <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  min.aln.len <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  disable.gapped <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  model.type <span class="op">=</span> <span class="st">"snp"</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  bound.low <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  bound.high <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  bc.err.freq <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-mismatch-testing">mismatch.testing<a class="anchor" aria-label="anchor" href="#arg-mismatch-testing"></a></dt>
<dd><p>(logical). Default: <code>mismatch.testing = FALSE</code>.</p></dd>


<dt id="arg-sample-list">sample.list<a class="anchor" aria-label="anchor" href="#arg-sample-list"></a></dt>
<dd><p>The default use all the samples in <code>f</code>,
usually, the stacks process_radtags output folder, here defaulted to
<code>f = "04_process_radtags"</code>, see <code>f</code> argument below.
When using <code>mismatch.testing = TRUE</code>, only one sample is allowed inside the folder specified in <code>f</code>
(e.g. choose one with a mean number of read, or MB, file size).
Default: <code>sample.list = NULL</code>.
Power outage? no problem, see details below.</p></dd>


<dt id="arg-project-info">project.info<a class="anchor" aria-label="anchor" href="#arg-project-info"></a></dt>
<dd><p>When using the stackr pipeline,
a project info file is created. This file will be modified inside this function.
The file is in the working directory (given the path or in the global environment).
If no <code>project.info</code> file is provided, the function first look in the
working directory for file(s) with "project.info" in it's name. If several files are found,
the latest one created is used.
Default: <code>project.info = NULL</code>.
Power outage? no problem, see details below.</p></dd>


<dt id="arg-f">f<a class="anchor" aria-label="anchor" href="#arg-f"></a></dt>
<dd><p>Input file path. Usually,
the stacks process_radtags output folder.
Default: <code>f = "04_process_radtags"</code>.</p></dd>


<dt id="arg-o">o<a class="anchor" aria-label="anchor" href="#arg-o"></a></dt>
<dd><p>Output path to write results.
Default: <code>o = "06_ustacks_2_gstacks"</code>.</p></dd>


<dt id="arg-m">m<a class="anchor" aria-label="anchor" href="#arg-m"></a></dt>
<dd><p>Minimum depth of coverage required to create a stack.
Default: <code>m = 3</code>.</p></dd>


<dt id="arg-m">M<a class="anchor" aria-label="anchor" href="#arg-m"></a></dt>
<dd><p>Maximum distance (in nucleotides) allowed between stacks.
Default: <code>M = 2</code>.</p></dd>


<dt id="arg-n">N<a class="anchor" aria-label="anchor" href="#arg-n"></a></dt>
<dd><p>Maximum distance allowed to align secondary reads to primary stacks.
Default: <code>N = M + 2</code>.</p></dd>


<dt id="arg-t">t<a class="anchor" aria-label="anchor" href="#arg-t"></a></dt>
<dd><p>Input file type.
Supported types: fasta, fastq, gzfasta, gzfastq, fq.gz, fastq.gz.
Default: <code>t = "guess"</code>.</p></dd>


<dt id="arg-r">R<a class="anchor" aria-label="anchor" href="#arg-r"></a></dt>
<dd><p>Retain unused reads. Default: <code>R = FALSE</code>.</p></dd>


<dt id="arg-h">H<a class="anchor" aria-label="anchor" href="#arg-h"></a></dt>
<dd><p>Disable calling haplotypes from secondary reads.
Default: <code>H = TRUE</code>.</p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>Enable parallel execution with num_threads threads.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></dd>


<dt id="arg-h">h<a class="anchor" aria-label="anchor" href="#arg-h"></a></dt>
<dd><p>Display this help messsage. Default: <code>h = FALSE</code>.</p></dd>


<dt id="arg-d">d<a class="anchor" aria-label="anchor" href="#arg-d"></a></dt>
<dd><p>Enable the Deleveraging algorithm, used for resolving over merged tags.
Default: <code>d = TRUE</code>.</p></dd>


<dt id="arg-keep-high-cov">keep.high.cov<a class="anchor" aria-label="anchor" href="#arg-keep-high-cov"></a></dt>
<dd><p>Disable the algorithm that removes highly-repetitive stacks and nearby errors.
Default: <code>keep.high.cov = FALSE</code>.</p></dd>


<dt id="arg-high-cov-thres">high.cov.thres<a class="anchor" aria-label="anchor" href="#arg-high-cov-thres"></a></dt>
<dd><p>(double) Highly-repetitive stacks threshold,
in standard deviation units.
Default: <code>high.cov.thres = 3.0</code>.</p></dd>


<dt id="arg-max-locus-stacks">max.locus.stacks<a class="anchor" aria-label="anchor" href="#arg-max-locus-stacks"></a></dt>
<dd><p>Maximum number of stacks at a single de novo locus.
Default: <code>max.locus.stacks = 3</code>.</p></dd>


<dt id="arg-k-len">k.len<a class="anchor" aria-label="anchor" href="#arg-k-len"></a></dt>
<dd><p>Specify k-mer size for matching between alleles and loci.
Default: <code>k.len = NULL</code>.</p></dd>


<dt id="arg-max-gaps">max.gaps<a class="anchor" aria-label="anchor" href="#arg-max-gaps"></a></dt>
<dd><p>Number of gaps allowed between stacks before merging.
Default: <code>max.gaps = 2</code>.</p></dd>


<dt id="arg-min-aln-len">min.aln.len<a class="anchor" aria-label="anchor" href="#arg-min-aln-len"></a></dt>
<dd><p>Minimum length of aligned sequence in a gapped alignment.
Default: <code>min.aln.len = 0.8</code>.</p></dd>


<dt id="arg-disable-gapped">disable.gapped<a class="anchor" aria-label="anchor" href="#arg-disable-gapped"></a></dt>
<dd><p>(logical) do not preform gapped alignments between stacks
(default: gapped alignements enabled).
Default: <code>disable.gapped = FALSE</code>.</p></dd>


<dt id="arg-model-type">model.type<a class="anchor" aria-label="anchor" href="#arg-model-type"></a></dt>
<dd><p>Either 'snp' (default), 'bounded', or 'fixed'.
Default: <code>model.type = "snp"</code>.</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>For the SNP or Bounded SNP model,
Chi square significance level required to call
a heterozygote or homozygote, either 0.1, 0.05.
Default: <code>alpha = 0.05</code>.</p></dd>


<dt id="arg-bound-low">bound.low<a class="anchor" aria-label="anchor" href="#arg-bound-low"></a></dt>
<dd><p>For the bounded SNP model, lower bound for epsilon,
the error rate, between 0 and 1.0.
Default: <code>bound.low = 0</code>.</p></dd>


<dt id="arg-bound-high">bound.high<a class="anchor" aria-label="anchor" href="#arg-bound-high"></a></dt>
<dd><p>For the bounded SNP model, upper bound for epsilon,
the error rate, between 0 and 1.0.
Default: <code>bound.high = 0.2</code>.</p></dd>


<dt id="arg-bc-err-freq">bc.err.freq<a class="anchor" aria-label="anchor" href="#arg-bc-err-freq"></a></dt>
<dd><p>For the fixed model, specify the barcode error frequency, between 0 and 1.0.
Default: <code>bc.err.freq = NULL</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p><a href="http://catchenlab.life.illinois.edu/stacks/comp/ustacks.php" class="external-link">ustacks</a>
returns 4 files per samples: <code>.snps.tsv.gz</code>, <code>.tags.tsv.gz</code>,
<code>.alleles.tsv.gz</code>,  <code>.models.tsv.gz</code>. In the global environment,
the function returns a project info file (updated if one was provided) and
a summary of ustacks for each samples.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><code>-i</code> the unique integer ID to identify the sample (SQL ID),
is taken from the project info file. If no project info file is provided,
the id is created sequentially from the sample files. This id will be written
in the project info file (if no file is found or given, a new file is created in the working directory).</p>
<p><strong>Power outage? No problem:</strong></p>
<p>Restart the function as it was. After the re-start the project info file created automatically during the previous run
will be used. This ensure: i) that the unique SQL ids are not duplicated and
ii) that ustacks can start at the sample is was assembling before the outage.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Catchen JM, Amores A, Hohenlohe PA et al. (2011)
Stacks: Building and Genotyping Loci De Novo From Short-Read Sequences.
G3, 1, 171-182.</p>
<p>Catchen JM, Hohenlohe PA, Bassham S, Amores A, Cresko WA (2013)
Stacks: an analysis tool set for population genomics.
Molecular Ecology, 22, 3124-3140.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="http://catchenlab.life.illinois.edu/stacks/comp/ustacks.php" class="external-link">ustacks</a></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Thierry Gosselin <a href="mailto:thierrygosselin@icloud.com">thierrygosselin@icloud.com</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>

    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

