<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Run STACKS cstacks module — run_cstacks • stackr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Run STACKS cstacks module — run_cstacks"><meta property="og:description" content="Run STACKS
cstacks
module inside R! The function runs a summary of the log file automatically
at the end (summary_cstacks). In the event of a power outage,
computer or cluster crash, just re-run the function. The function will start
over from the last catalog generated."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stackr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/stackr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/stackr" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run STACKS cstacks module</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/stackr/blob/HEAD/R/run_cstacks.R" class="external-link"><code>R/run_cstacks.R</code></a></small>
    <div class="hidden name"><code>run_cstacks.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run <a href="http://catchenlab.life.illinois.edu/stacks/" class="external-link">STACKS</a>
<a href="http://catchenlab.life.illinois.edu/stacks/comp/cstacks.php" class="external-link">cstacks</a>
module inside R! The function runs a summary of the log file automatically
at the end (<code><a href="summary_cstacks.html">summary_cstacks</a></code>). In the event of a power outage,
computer or cluster crash, just re-run the function. The function will start
over from the last catalog generated.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_cstacks</span><span class="op">(</span></span>
<span>  P <span class="op">=</span> <span class="st">"06_ustacks_2_gstacks"</span>,</span>
<span>  o <span class="op">=</span> <span class="st">"06_ustacks_2_gstacks"</span>,</span>
<span>  M <span class="op">=</span> <span class="st">"02_project_info/population.map.catalog.tsv"</span>,</span>
<span>  catalog.path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>  max.gaps <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  min.aln.len <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  disable.gapped <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  k.len <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  report.mmatches <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  split.catalog <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-p">P<a class="anchor" aria-label="anchor" href="#arg-p"></a></dt>
<dd><p>path to the directory containing STACKS files.
Default: <code>P = "06_ustacks_2_gstacks"</code>.
Inside the folder <code>06_ustacks_2_gstacks</code>, you should have:</p><ul><li><p><strong>4 files for each samples:</strong> The sample name is the prefix of
  the files ending with:
<code>.alleles.tsv.gz, .models.tsv.gz, .snps.tsv.gz, .tags.tsv.gz</code>.
Those files are created in the
<a href="http://catchenlab.life.illinois.edu/stacks/comp/ustacks.php" class="external-link">ustacks</a>
module.</p></li>
</ul></dd>


<dt id="arg-o">o<a class="anchor" aria-label="anchor" href="#arg-o"></a></dt>
<dd><p>Output path to write catalog.
Default: <code>o = "06_ustacks_2_gstacks"</code></p></dd>


<dt id="arg-m">M<a class="anchor" aria-label="anchor" href="#arg-m"></a></dt>
<dd><p>path to a population map file (Required when P is used).
Default: <code>M = "02_project_info/population.map.catalog.tsv"</code>.</p></dd>


<dt id="arg-catalog-path">catalog.path<a class="anchor" aria-label="anchor" href="#arg-catalog-path"></a></dt>
<dd><p>This is for the "Catalog editing" part in cstacks where
you can provide the path to an existing catalog.
cstacks will add data to this existing catalog.
With default: <code>catalog.path = NULL</code> or with a supplied path, the function
The function scan automatically for the presence of a catalog inside the input folder.
If none is found, a new catalog is created.
If your catalog is not in the input folder, supply a path here.
e.g. <code>catalog.path = ~/catalog_folder</code>, the catalog files are inside the
P folder along the samples files and detected automatically.
If a catalog is detected in the input folder,
the samples in the <code>sample.list</code> argument
will be added in this catalog. The catalog is made of 3 files:
<code>catalog.alleles.tsv.gz, catalog.snps.tsv.gz, catalog.tags.tsv.gz</code></p></dd>


<dt id="arg-n">n<a class="anchor" aria-label="anchor" href="#arg-n"></a></dt>
<dd><p>number of mismatches allowed between sample loci when build the catalog.
Default: <code>n = 1</code></p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>Enable parallel execution with num_threads threads.
Default: <code>parallel.core = parallel::detectCores() - 1</code></p></dd>


<dt id="arg-max-gaps">max.gaps<a class="anchor" aria-label="anchor" href="#arg-max-gaps"></a></dt>
<dd><p>The number of gaps allowed between stacks before merging.
Default: <code>max.gaps = 2</code></p></dd>


<dt id="arg-min-aln-len">min.aln.len<a class="anchor" aria-label="anchor" href="#arg-min-aln-len"></a></dt>
<dd><p>The minimum length of aligned sequence in a gapped
alignment.
Default: <code>min.aln.len = 0.8</code></p></dd>


<dt id="arg-disable-gapped">disable.gapped<a class="anchor" aria-label="anchor" href="#arg-disable-gapped"></a></dt>
<dd><p>Disable gapped alignments between stacks.
Default: <code>disable.gapped = FALSE</code> (use gapped alignments).</p></dd>


<dt id="arg-k-len">k.len<a class="anchor" aria-label="anchor" href="#arg-k-len"></a></dt>
<dd><p>Specify k-mer size for matching between between catalog loci
(automatically calculated by default).
Advice: don't modify.
Default: <code>k.len = NULL</code></p></dd>


<dt id="arg-report-mmatches">report.mmatches<a class="anchor" aria-label="anchor" href="#arg-report-mmatches"></a></dt>
<dd><p>Report query loci that match more than one catalog locus.
Advice: don't modify.
Default: <code>report.mmatches = FALSE</code></p></dd>


<dt id="arg-split-catalog">split.catalog<a class="anchor" aria-label="anchor" href="#arg-split-catalog"></a></dt>
<dd><p>(integer) In how many samples you want to split the
catalog population map. This allows to have a backup catalog every
<code>split.catalog</code> samples. Their is obviously a trade-off between the
integer use here, the time to initialize an existing catalog and
re-starting from zero if everything crash.
Default: <code>split.catalog = 20</code>. Very useful on a personal computer or
university computer cluster....</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p><a href="http://catchenlab.life.illinois.edu/stacks/comp/sstacks.php" class="external-link">sstacks</a>
returns a <code>.matches.tsv.gz file for each sample</code></p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><strong>Computer or server problem during the cstacks ?</strong> Look
in the log file to see which individuals remains to be included. Create a
new list of individuals to include and use the catalog.path argument to point
to the catalog created before the problem.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Catchen JM, Amores A, Hohenlohe PA et al. (2011)
Stacks: Building and Genotyping Loci De Novo From Short-Read Sequences.
G3, 1, 171-182.</p>
<p>Catchen JM, Hohenlohe PA, Bassham S, Amores A, Cresko WA (2013)
Stacks: an analysis tool set for population genomics.
Molecular Ecology, 22, 3124-3140.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="http://catchenlab.life.illinois.edu/stacks/comp/sstacks.php" class="external-link">sstacks</a></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>

    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

