<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Run STACKS sstacks module — run_sstacks • stackr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Run STACKS sstacks module — run_sstacks"><meta property="og:description" content="Run STACKS
sstacks
module inside R!
Inside the P folder (where the ustacks and the cstacks files are),
you should have:
3 Catalog files: the files created in
cstacks
and usually looking like this:
batch_1.catalog.alleles.tsv.gz,
batch_1.catalog.snps.tsv.gz,
batch_1.catalog.tags.tsv.gz
3 files for each samples: The sample name is the prefix of
  the files ending with:
.alleles.tsv.gz, .snps.tsv.gz, .tags.tsv.gz.
Those files are created in the
ustacks
module.

"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stackr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/stackr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/stackr" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run STACKS sstacks module</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/stackr/blob/HEAD/R/run_sstacks.R" class="external-link"><code>R/run_sstacks.R</code></a></small>
    <div class="hidden name"><code>run_sstacks.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run <a href="http://catchenlab.life.illinois.edu/stacks/" class="external-link">STACKS</a>
<a href="http://catchenlab.life.illinois.edu/stacks/comp/sstacks.php" class="external-link">sstacks</a>
module inside R!
Inside the <code>P</code> folder (where the <a href="http://catchenlab.life.illinois.edu/stacks/comp/ustacks.php" class="external-link">ustacks</a> and the <a href="http://catchenlab.life.illinois.edu/stacks/comp/cstacks.php" class="external-link">cstacks</a> files are),
you should have:</p><ul><li><p><strong>3 Catalog files:</strong> the files created in
<a href="http://catchenlab.life.illinois.edu/stacks/comp/cstacks.php" class="external-link">cstacks</a>
and usually looking like this:
<code>batch_1.catalog.alleles.tsv.gz,
batch_1.catalog.snps.tsv.gz,
batch_1.catalog.tags.tsv.gz</code></p></li>
<li><p><strong>3 files for each samples:</strong> The sample name is the prefix of
  the files ending with:
<code>.alleles.tsv.gz, .snps.tsv.gz, .tags.tsv.gz</code>.
Those files are created in the
<a href="http://catchenlab.life.illinois.edu/stacks/comp/ustacks.php" class="external-link">ustacks</a>
module.</p></li>
</ul></div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_sstacks</span><span class="op">(</span></span>
<span>  P <span class="op">=</span> <span class="st">"06_ustacks_2_gstacks"</span>,</span>
<span>  M <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample.list <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  c <span class="op">=</span> <span class="st">"06_ustacks_2_gstacks"</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>  o <span class="op">=</span> <span class="st">"06_ustacks_2_gstacks"</span>,</span>
<span>  x <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  disable.gapped <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  v <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  h <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-p">P<a class="anchor" aria-label="anchor" href="#arg-p"></a></dt>
<dd><p>(character) Path to the directory containing STACKS files.
Contrary to <a href="http://catchenlab.life.illinois.edu/stacks/comp/sstacks.php" class="external-link">sstacks</a>,
this argument can be use with <code>sample.list</code> (the <code>s</code> in <a href="http://catchenlab.life.illinois.edu/stacks/comp/sstacks.php" class="external-link">sstacks</a>)
Default: <code>P = "06_ustacks_2_gstacks"</code>.</p></dd>


<dt id="arg-m">M<a class="anchor" aria-label="anchor" href="#arg-m"></a></dt>
<dd><p>(character) Path to a population map file from which to take sample
names (this argument won't work if <code>sample.list</code>).
Advice: don't modify the default (please see details).
Default: <code>M = NULL</code>.</p></dd>


<dt id="arg-sample-list">sample.list<a class="anchor" aria-label="anchor" href="#arg-sample-list"></a></dt>
<dd><p>This is for the <code>s</code> option in
<a href="http://catchenlab.life.illinois.edu/stacks/comp/sstacks.php" class="external-link">sstacks</a>.
<code>s: Filename prefix from which to load sample loci</code>.
Here, you have 2 choices: 1. you leave empty and let the function use the
default:<code>sample.list = NULL</code> which will scan for the files in the
<code>P</code> folder given above. 2. you supply a character string of the
samples. This could come from the <code>INDIVIDUALS_REP</code> column of the
project info file, e.g. <code>sample.list = project.info$INDIVIDUALS_REP</code>.
(Please see details).</p></dd>


<dt id="arg-c">c<a class="anchor" aria-label="anchor" href="#arg-c"></a></dt>
<dd><p>(character, path) Path to the catalog.
Default: <code>c = "06_ustacks_2_gstacks"</code></p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(Integer) Enable parallel execution with num_threads threads.
Default: <code>parallel.core = parallel::detectCores() - 1</code></p></dd>


<dt id="arg-o">o<a class="anchor" aria-label="anchor" href="#arg-o"></a></dt>
<dd><p>output path to write results.
Default: <code>o = "06_ustacks_2_gstacks"</code></p></dd>


<dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Don't verify haplotype of matching locus.
Default: <code>x = FALSE</code></p></dd>


<dt id="arg-disable-gapped">disable.gapped<a class="anchor" aria-label="anchor" href="#arg-disable-gapped"></a></dt>
<dd><p>Disable gapped alignments between stacks.
Default: <code>disable.gapped = FALSE</code> (use gapped alignments).</p></dd>


<dt id="arg-v">v<a class="anchor" aria-label="anchor" href="#arg-v"></a></dt>
<dd><p>Print program version.
Default: <code>v = FALSE</code></p></dd>


<dt id="arg-h">h<a class="anchor" aria-label="anchor" href="#arg-h"></a></dt>
<dd><p>Display this help messsage.
Default: <code>h = FALSE</code></p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p><a href="http://catchenlab.life.illinois.edu/stacks/comp/sstacks.php" class="external-link">sstacks</a>
returns a <code>.matches.tsv.gz file for each sample</code>.
If <code>lnl_dist = TRUE</code>, the function will also return a
summary of catalog loci log-likelihood.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><strong>Computer or server problem during the sstacks ?</strong>
Just launch again the same command, the function will start again, but only
with the unmatched samples!</p>
<p><strong>Some argument can't be use together:</strong></p>
<p>Don't use these arguments together: <code>M</code> and <code>sample.list/c</code>.
You can't add samples to an existing catalog using a population map.
If you want to match new samples to an existing catalog, very easy,
just add the samples in the folder and let <code>run_sstacks</code> function find
the samples that were not matched to the catalog.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Catchen JM, Amores A, Hohenlohe PA et al. (2011)
Stacks: Building and Genotyping Loci De Novo From Short-Read Sequences.
G3, 1, 171-182.</p>
<p>Catchen JM, Hohenlohe PA, Bassham S, Amores A, Cresko WA (2013)
Stacks: an analysis tool set for population genomics.
Molecular Ecology, 22, 3124-3140.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="http://catchenlab.life.illinois.edu/stacks/comp/sstacks.php" class="external-link">sstacks</a></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>

    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

