<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Run STACKS Version &gt;= 2.0 populations module — run_populations • stackr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Run STACKS Version &gt;= 2.0 populations module — run_populations"><meta property="og:description" content="Run STACKS
populations
module inside R!"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stackr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/stackr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thierrygosselin/stackr" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run STACKS Version &gt;= 2.0 populations module</h1>
    <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/stackr/blob/HEAD/R/run_populations.R" class="external-link"><code>R/run_populations.R</code></a></small>
    <div class="hidden name"><code>run_populations.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run <a href="http://catchenlab.life.illinois.edu/stacks/" class="external-link">STACKS</a>
<a href="http://catchenlab.life.illinois.edu/stacks/comp/populations.php" class="external-link">populations</a>
module inside R!</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_populations</span><span class="op">(</span></span>
<span>  P <span class="op">=</span> <span class="st">"06_ustacks_2_gstacks"</span>,</span>
<span>  V <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  O <span class="op">=</span> <span class="st">"07_populations"</span>,</span>
<span>  <span class="va">M</span>,</span>
<span>  parallel.core <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>  batch.size <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  p <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  r <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  R <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  H <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  min.maf <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min.mac <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  max.obs.het <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  write.single.snp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  write.random.snp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  B <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  W <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  e <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  merge.sites <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  merge.prune.lim <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  hwe <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fstats <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fst.correction <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p.value.cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  k <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  smooth.fstats <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  smooth.popstats <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sigma <span class="op">=</span> <span class="fl">150000</span>,</span>
<span>  bootstrap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  N <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  bootstrap.pifis <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  bootstrap.fst <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  bootstrap.div <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  bootstrap.phist <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  bootstrap.wl <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ordered.export <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fasta.samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  fasta.samples_raw <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fasta.loci <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  vcf <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  genepop <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  structure <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  radpainter <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  phase <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  fastphase <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plink <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  hzar <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  phylip <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  phylip.var <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  phylip.var.all <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  treemix <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  no.hap.exports <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  h <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  v <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  log.fst.comp <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-p">P<a class="anchor" aria-label="anchor" href="#arg-p"></a></dt>
<dd><p>(path, character) Path to the directory containing all the STACKS files.
Default: <code>P = "06_ustacks_2_gstacks"</code>.</p></dd>


<dt id="arg-v">V<a class="anchor" aria-label="anchor" href="#arg-v"></a></dt>
<dd><p>(character) Path to an input VCF file. When this module is used to
filter an existing vcf file.
Default: <code>V = NULL</code>.</p></dd>


<dt id="arg-o">O<a class="anchor" aria-label="anchor" href="#arg-o"></a></dt>
<dd><p>(character) Path to a directory where to write the output files.
With default: <code>O = "07_populations"</code>, the function creates a folder inside
<code>07_populations</code>, with date and time appended to <code>populations_</code>.</p></dd>


<dt id="arg-m">M<a class="anchor" aria-label="anchor" href="#arg-m"></a></dt>
<dd><p>path to a population map file. The format is a tab-separated file,
with first column containing sample name and second column population id.
No heather (column name).
e.g. <code>M = "02_project_info/population.map.catalog.tsv"</code></p></dd>


<dt id="arg-parallel-core">parallel.core<a class="anchor" aria-label="anchor" href="#arg-parallel-core"></a></dt>
<dd><p>(integer) enable parallel execution with num_threads threads.
Default: <code>parallel.core = parallel::detectCores() - 1</code></p></dd>


<dt id="arg-batch-size">batch.size<a class="anchor" aria-label="anchor" href="#arg-batch-size"></a></dt>
<dd><p>(integer) The number of loci (de novo mode) or
chromosome (reference mode), to process in a batch.
Increase to speed analysis, uses more memory, decrease to save memory).
Default in de novo mode (loci/batch): <code>batch.size = 10000</code>.
Default in reference mode (chromosome/batch): <code>batch.size = 1</code>.</p></dd>


<dt id="arg-p">p<a class="anchor" aria-label="anchor" href="#arg-p"></a></dt>
<dd><p>(integer) Minimum number of populations a locus must be present in to process a locus.
Default: <code>p = 1</code>.</p></dd>


<dt id="arg-r">r<a class="anchor" aria-label="anchor" href="#arg-r"></a></dt>
<dd><p>(double) Minimum percentage of individuals in a population required to process a locus for that population.
Default: <code>r = 0.3</code>.</p></dd>


<dt id="arg-r">R<a class="anchor" aria-label="anchor" href="#arg-r"></a></dt>
<dd><p>(double) Minimum percentage of individuals across populations required to process a locus.
Default: <code>r = 0.3</code>.</p></dd>


<dt id="arg-h">H<a class="anchor" aria-label="anchor" href="#arg-h"></a></dt>
<dd><p>(logical) Apply the above filters haplotype wise
(unshared SNPs will be pruned to reduce haplotype-wise missing data).
Default: <code>H = TRUE</code>.</p></dd>


<dt id="arg-min-maf">min.maf<a class="anchor" aria-label="anchor" href="#arg-min-maf"></a></dt>
<dd><p>(double) Specify a minimum minor allele frequency required to process a nucleotide site at a locus (0 &lt; min.maf &lt; 0.5).
Default: <code>min.maf = NULL</code>. Recommendation: use my other R package RADIATOR to filter data based on MAF.</p></dd>


<dt id="arg-min-mac">min.mac<a class="anchor" aria-label="anchor" href="#arg-min-mac"></a></dt>
<dd><p>(integer) Specify a minimum minor allele count required to process a nucleotide site at a locus.
Default: <code>min.mac = NULL</code>.</p></dd>


<dt id="arg-max-obs-het">max.obs.het<a class="anchor" aria-label="anchor" href="#arg-max-obs-het"></a></dt>
<dd><p>(double) Specify a maximum observed heterozygosity required to process a nucleotide site at a locus.
Default: <code>max.obs.het = NULL</code>.  Recommendation: use my other R package RADIATOR to filter data based on heterozygosity.</p></dd>


<dt id="arg-write-single-snp">write.single.snp<a class="anchor" aria-label="anchor" href="#arg-write-single-snp"></a></dt>
<dd><p>(logical) Restrict data analysis to only the first SNP per locus (implies –no-haps).
Default: <code>write.single.snp = FALSE</code>.</p></dd>


<dt id="arg-write-random-snp">write.random.snp<a class="anchor" aria-label="anchor" href="#arg-write-random-snp"></a></dt>
<dd><p>(logical) Restrict data analysis to one random SNP per locus (implies –no-haps).
Default: <code>write.random.snp = FALSE</code>.</p></dd>


<dt id="arg-b">B<a class="anchor" aria-label="anchor" href="#arg-b"></a></dt>
<dd><p>(character) Path to a file containing Blacklisted markers to be excluded from the export.
Default: <code>B = NULL</code>.</p></dd>


<dt id="arg-w">W<a class="anchor" aria-label="anchor" href="#arg-w"></a></dt>
<dd><p>(character) Path to a file containing Whitelisted markers to include in the export.
Default: <code>W = NULL</code>.</p></dd>


<dt id="arg-e">e<a class="anchor" aria-label="anchor" href="#arg-e"></a></dt>
<dd><p>(character) Restriction enzyme name.
Default: <code>e = NULL</code>.</p></dd>


<dt id="arg-merge-sites">merge.sites<a class="anchor" aria-label="anchor" href="#arg-merge-sites"></a></dt>
<dd><p>(logical) Merge loci that were produced from the same restriction enzyme cutsite (requires reference-aligned data).
Default: <code>merge.sites = FALSE</code>.</p></dd>


<dt id="arg-merge-prune-lim">merge.prune.lim<a class="anchor" aria-label="anchor" href="#arg-merge-prune-lim"></a></dt>
<dd><p>(integer) When merging adjacent loci, if at least X<!-- % samples posses both loci prune the remaining samples out of the analysis. -->
Default: <code>merge.prune.lim = NULL</code>.</p></dd>


<dt id="arg-hwe">hwe<a class="anchor" aria-label="anchor" href="#arg-hwe"></a></dt>
<dd><p>(logical) Calculate divergence from Hardy-Weinberg equilibrium
using the exact test at the SNP level and Guo and Thompson MCMC algorithm at
the haplotype level.
Default: <code>hwe = FALSE</code>.</p></dd>


<dt id="arg-fstats">fstats<a class="anchor" aria-label="anchor" href="#arg-fstats"></a></dt>
<dd><p>(logical) Enable SNP and haplotype-based F statistics.
Default: <code>fstats = FALSE</code>.</p></dd>


<dt id="arg-fst-correction">fst.correction<a class="anchor" aria-label="anchor" href="#arg-fst-correction"></a></dt>
<dd><p>(character) Specify a correction to be applied to Fst values: 'p_value', 'bonferroni_win', or 'bonferroni_gen'.
Default: <code>fst.correction = NULL</code>.</p></dd>


<dt id="arg-p-value-cutoff">p.value.cutoff<a class="anchor" aria-label="anchor" href="#arg-p-value-cutoff"></a></dt>
<dd><p>(double) maximum p-value to keep an Fst measurement.
Also used as base for Bonferroni correction.
Default: <code>p.value.cutoff = 0.05</code>.</p></dd>


<dt id="arg-k">k<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>(logical) Enable kernel-smoothed Pi, Fis, Fst, Fst', and Phi_st calculations.
Default: <code>k = FALSE</code>.</p></dd>


<dt id="arg-smooth-fstats">smooth.fstats<a class="anchor" aria-label="anchor" href="#arg-smooth-fstats"></a></dt>
<dd><p>(logical) Enable kernel-smoothed Fst, Fst', and Phi_st calculations.
Default: <code>smooth.fstats = FALSE</code>.</p></dd>


<dt id="arg-smooth-popstats">smooth.popstats<a class="anchor" aria-label="anchor" href="#arg-smooth-popstats"></a></dt>
<dd><p>(logical) Enable kernel-smoothed Pi and Fis calculations.
Default: <code>smooth.popstats = FALSE</code>.</p></dd>


<dt id="arg-sigma">sigma<a class="anchor" aria-label="anchor" href="#arg-sigma"></a></dt>
<dd><p>(integer) Standard deviation of the kernel smoothing weight distribution.
Default: <code>sigma = 150000</code> (150kb).</p></dd>


<dt id="arg-bootstrap">bootstrap<a class="anchor" aria-label="anchor" href="#arg-bootstrap"></a></dt>
<dd><p>(logical) Turn on boostrap resampling for all smoothed statistics.
Default: <code>bootstrap = FALSE</code>.</p></dd>


<dt id="arg-n">N<a class="anchor" aria-label="anchor" href="#arg-n"></a></dt>
<dd><p>(integer) Number of bootstrap resamplings to calculate.
Default: <code>N = 100</code>.</p></dd>


<dt id="arg-bootstrap-pifis">bootstrap.pifis<a class="anchor" aria-label="anchor" href="#arg-bootstrap-pifis"></a></dt>
<dd><p>(logical) Turn on boostrap resampling for smoothed SNP-based Pi and Fis calculations.
Default: <code>bootstrap.pifis = FALSE</code>.</p></dd>


<dt id="arg-bootstrap-fst">bootstrap.fst<a class="anchor" aria-label="anchor" href="#arg-bootstrap-fst"></a></dt>
<dd><p>(logical) Turn on boostrap resampling for smoothed Fst calculations based on pairwise population comparison of SNPs.
Default: <code>bootstrap.fst = FALSE</code>.</p></dd>


<dt id="arg-bootstrap-div">bootstrap.div<a class="anchor" aria-label="anchor" href="#arg-bootstrap-div"></a></dt>
<dd><p>(logical) Turn on boostrap resampling for smoothed haplotype diveristy and gene diversity calculations based on haplotypes.
Default: <code>bootstrap.div = FALSE</code>.</p></dd>


<dt id="arg-bootstrap-phist">bootstrap.phist<a class="anchor" aria-label="anchor" href="#arg-bootstrap-phist"></a></dt>
<dd><p>(logical) Turn on boostrap resampling for smoothed Phi_st calculations based on haplotypes.
Default: <code>bootstrap.phist = FALSE</code>.</p></dd>


<dt id="arg-bootstrap-wl">bootstrap.wl<a class="anchor" aria-label="anchor" href="#arg-bootstrap-wl"></a></dt>
<dd><p>(character) Path to a whitelist file. Only use bootstrap loci contained in this whitelist.
Default: <code>bootstrap.wl = NULL</code>.</p></dd>


<dt id="arg-ordered-export">ordered.export<a class="anchor" aria-label="anchor" href="#arg-ordered-export"></a></dt>
<dd><p>(logical) If data is reference aligned, exports will be ordered; only a single representative of each overlapping site.
Default: <code>ordered.export = FALSE</code>.</p></dd>


<dt id="arg-fasta-samples">fasta.samples<a class="anchor" aria-label="anchor" href="#arg-fasta-samples"></a></dt>
<dd><p>(logical) Output the sequences of the two haplotypes of each (diploid) sample, for each locus, in FASTA format.
stackr default is different than stacks default because I think having this info helps to identify problems.
Default: <code>fasta.samples = TRUE</code>.</p></dd>


<dt id="arg-fasta-samples-raw">fasta.samples_raw<a class="anchor" aria-label="anchor" href="#arg-fasta-samples-raw"></a></dt>
<dd><p>(logical) Output all haplotypes observed in each sample, for each locus, in FASTA format.
Default: <code>fasta.samples_raw = FALSE</code>.</p></dd>


<dt id="arg-fasta-loci">fasta.loci<a class="anchor" aria-label="anchor" href="#arg-fasta-loci"></a></dt>
<dd><p>(logical) Output consensus sequences of all loci, in FASTA format.
stackr default is different than stacks default because I think having this info helps to identify problems.
Default: <code>fasta.loci = TRUE</code>.</p></dd>


<dt id="arg-vcf">vcf<a class="anchor" aria-label="anchor" href="#arg-vcf"></a></dt>
<dd><p>(logical) Output SNPs in Variant Call Format (VCF).
Default: <code>vcf = TRUE</code>.</p></dd>


<dt id="arg-genepop">genepop<a class="anchor" aria-label="anchor" href="#arg-genepop"></a></dt>
<dd><p>(logical) Output results in GenePop format.
Default: <code>genepop = FALSE</code>.</p></dd>


<dt id="arg-structure">structure<a class="anchor" aria-label="anchor" href="#arg-structure"></a></dt>
<dd><p>(logical) Output results in Structure format.
Default: <code>structure = FALSE</code>.</p></dd>


<dt id="arg-radpainter">radpainter<a class="anchor" aria-label="anchor" href="#arg-radpainter"></a></dt>
<dd><p>(logical) Output results results in fineRADstructure/RADpainter format.
Default: <code>radpainter = FALSE</code>.</p></dd>


<dt id="arg-phase">phase<a class="anchor" aria-label="anchor" href="#arg-phase"></a></dt>
<dd><p>(logical) Output genotypes in PHASE format.
Default: <code>phase = FALSE</code>.</p></dd>


<dt id="arg-fastphase">fastphase<a class="anchor" aria-label="anchor" href="#arg-fastphase"></a></dt>
<dd><p>(logical) Output genotypes in fastPHASE format.
Default: <code>fastphase = FALSE</code>.</p></dd>


<dt id="arg-plink">plink<a class="anchor" aria-label="anchor" href="#arg-plink"></a></dt>
<dd><p>(logical) Output genotypes in PLINK format.
Default: <code>plink = FALSE</code>.</p></dd>


<dt id="arg-hzar">hzar<a class="anchor" aria-label="anchor" href="#arg-hzar"></a></dt>
<dd><p>(logical) Output genotypes in Hybrid Zone Analysis using R (HZAR) format.
Default: <code>hzar = FALSE</code>.</p></dd>


<dt id="arg-phylip">phylip<a class="anchor" aria-label="anchor" href="#arg-phylip"></a></dt>
<dd><p>(logical) Output nucleotides that are fixed-within, and variant among populations in Phylip format for phylogenetic tree construction.
Default: <code>phylip = FALSE</code>.</p></dd>


<dt id="arg-phylip-var">phylip.var<a class="anchor" aria-label="anchor" href="#arg-phylip-var"></a></dt>
<dd><p>(logical) Include variable sites in the phylip output encoded using IUPAC notation.
Default: <code>phylip.var = FALSE</code>.</p></dd>


<dt id="arg-phylip-var-all">phylip.var.all<a class="anchor" aria-label="anchor" href="#arg-phylip-var-all"></a></dt>
<dd><p>(logical) Include all sequence as well as variable sites in the phylip output encoded using IUPAC notation.
Default: <code>phylip.var.all = FALSE</code>.</p></dd>


<dt id="arg-treemix">treemix<a class="anchor" aria-label="anchor" href="#arg-treemix"></a></dt>
<dd><p>(logical) Output SNPs in a format useable for the TreeMix program (Pickrell and Pritchard).
Default: <code>treemix = FALSE</code>.</p></dd>


<dt id="arg-no-hap-exports">no.hap.exports<a class="anchor" aria-label="anchor" href="#arg-no-hap-exports"></a></dt>
<dd><p>(logical) Omit haplotype outputs.
Default: <code>no.hap.exports = FALSE</code>.</p></dd>


<dt id="arg-h">h<a class="anchor" aria-label="anchor" href="#arg-h"></a></dt>
<dd><p>Display this help messsage.
Default: <code>h = FALSE</code></p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>turn on additional logging.
Default: <code>verbose = FALSE</code></p></dd>


<dt id="arg-v">v<a class="anchor" aria-label="anchor" href="#arg-v"></a></dt>
<dd><p>print program version.
Default: <code>v = FALSE</code></p></dd>


<dt id="arg-log-fst-comp">log.fst.comp<a class="anchor" aria-label="anchor" href="#arg-log-fst-comp"></a></dt>
<dd><p>(logical) Log components of Fst/Phi_st calculations to a file.
Default: <code>log.fst.comp = FALSE</code></p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p><a href="http://catchenlab.life.illinois.edu/stacks/comp/populations.php" class="external-link">populations</a>
returns different output depending on arguments selected.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Catchen JM, Amores A, Hohenlohe PA et al. (2011)
Stacks: Building and Genotyping Loci De Novo From Short-Read Sequences.
G3, 1, 171-182.</p>
<p>Catchen JM, Hohenlohe PA, Bassham S, Amores A, Cresko WA (2013)
Stacks: an analysis tool set for population genomics.
Molecular Ecology, 22, 3124-3140.</p>
<p>Guo SW, Thompson EA (1992)
Performing the exact test of Hardy-Weinberg proportion for multiple alleles.
Biometrics, 48, 361-372.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="http://catchenlab.life.illinois.edu/stacks/comp/populations.php" class="external-link">populations</a>
<a href="http://catchenlab.life.illinois.edu/stacks/" class="external-link">STACKS Version 2.0b</a></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">pop</span> <span class="op">&lt;-</span> <span class="fu">stackr</span><span class="fu">::</span><span class="fu">run_populations</span><span class="op">(</span>M <span class="op">=</span> <span class="st">"population.map.tsv"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

